<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Saikat Banerjee">
<meta name="dcterms.date" content="2023-05-12">

<title>NPD Network - Preprocess NPD summary statistics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">NPD Network</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../notebooks/index.html" rel="" target="">
 <span class="menu-text">Notebooks</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../meetings/index.html" rel="" target="">
 <span class="menu-text">Meetings</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../resources/index.html" rel="" target="">
 <span class="menu-text">Resources</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../reference/index.html" rel="" target="">
 <span class="menu-text">Reference</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Contents</h2>
   
  <ul>
  <li><a href="#read" id="toc-read" class="nav-link active" data-scroll-target="#read">Read</a></li>
  <li><a href="#clean" id="toc-clean" class="nav-link" data-scroll-target="#clean">Clean</a></li>
  <li><a href="#data-matrix" id="toc-data-matrix" class="nav-link" data-scroll-target="#data-matrix">Data Matrix</a></li>
  <li><a href="#remove-studies-with-unreliable-standard-errors" id="toc-remove-studies-with-unreliable-standard-errors" class="nav-link" data-scroll-target="#remove-studies-with-unreliable-standard-errors">Remove studies with unreliable standard errors</a></li>
  <li><a href="#remove-studies-with-unreliable-z-scores" id="toc-remove-studies-with-unreliable-z-scores" class="nav-link" data-scroll-target="#remove-studies-with-unreliable-z-scores">Remove studies with unreliable z-scores</a></li>
  <li><a href="#flip-signs" id="toc-flip-signs" class="nav-link" data-scroll-target="#flip-signs">Flip Signs</a></li>
  <li><a href="#save-pickle" id="toc-save-pickle" class="nav-link" data-scroll-target="#save-pickle">Save Pickle</a></li>
  <li><a href="#save-snp-information" id="toc-save-snp-information" class="nav-link" data-scroll-target="#save-snp-information">Save SNP information</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.dev/banskt/npd-notes/blob/main/notebooks/explore/2023-05-12-preprocess.ipynb" class="toc-action">Edit this page</a></p><p><a href="https://github.com/banskt/npd-notes/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Preprocess NPD summary statistics</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Saikat Banerjee </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">May 12, 2023</p>
    </div>
  </div>
  
    
  </div>
  
<div>
  <div class="abstract">
    <div class="abstract-title">Abstract</div>
    We read the summary statistics for association of variants with NPD phenotypes in many GWAS and perform some basic filtering to obtain the data matrix for exploring the different methods.
  </div>
</div>

</header>

<div class="cell" data-execution_count="1">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy.stats <span class="im">as</span> sc_stats</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> collections</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pymir <span class="im">import</span> mpl_stylesheet</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pymir <span class="im">import</span> mpl_utils</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>mpl_stylesheet.banskt_presentation(splinecolor <span class="op">=</span> <span class="st">'black'</span>, dpi <span class="op">=</span> <span class="dv">120</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="read" class="level1">
<h1>Read</h1>
<p>Summary statistics for association of selected SNPs with multiple diseases. SNPs which had significant association with at least one phenotype were selected in a previous preprocessing step, details in <a href="https://github.com/daklab/npd_sumstats">this repository</a>. Next, the summary statistics of those SNPs for all available phenotypes were included from the respective GWAS.</p>
<div class="cell" data-execution_count="2">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>data_dir <span class="op">=</span> <span class="st">"../data"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>studies  <span class="op">=</span> [<span class="st">'gtex'</span>, <span class="st">'pgc'</span>, <span class="st">'ieu'</span>]</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>assoc_file <span class="op">=</span> {}</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> s <span class="kw">in</span> studies:</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>    assoc_file[s] <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>data_dir<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>s<span class="sc">}</span><span class="ss">_assoc_sb.txt.gz"</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>trait_df_filename <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>data_dir<span class="sc">}</span><span class="ss">/trait_meta.csv"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-execution_count="3">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>assoc <span class="op">=</span> {}</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>req_cols <span class="op">=</span> {</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">'SNP'</span> : <span class="st">'string'</span>,</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">'CHR'</span> : <span class="st">'string'</span>,</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">'BP'</span>  : <span class="st">'Int64'</span>, <span class="co"># use Pandas Int64 to handle NA values</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">'A1'</span>  : <span class="st">'string'</span>,</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">'A2'</span>  : <span class="st">'string'</span>,</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">'Z'</span>   : np.float64,</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">'P'</span>   : np.float64,</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">'BETA'</span>: np.float64,</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">'SE'</span>  : np.float64,</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">'ID'</span>  : <span class="st">'string'</span>,</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">'TRAIT'</span>: <span class="st">'string'</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> s <span class="kw">in</span> studies:</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span> (<span class="ss">f"Read summary statistics for </span><span class="sc">{</span>s<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>    header <span class="op">=</span> pd.read_csv(assoc_file[s], nrows<span class="op">=</span><span class="dv">0</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>    use_cols   <span class="op">=</span> [x <span class="cf">for</span> x <span class="kw">in</span> req_cols.keys() <span class="cf">if</span> x <span class="kw">in</span> header.columns]</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>    use_dtypes <span class="op">=</span> {k:v <span class="cf">for</span> k, v <span class="kw">in</span> req_cols.items() <span class="cf">if</span> k <span class="kw">in</span> use_cols}</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>    assoc[s] <span class="op">=</span> pd.read_csv(assoc_file[s], sep <span class="op">=</span> <span class="st">'</span><span class="ch">\t</span><span class="st">'</span>,</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>                           usecols <span class="op">=</span> use_cols,</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>                           dtype <span class="op">=</span> use_dtypes</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>                          )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Read summary statistics for gtex
Read summary statistics for pgc
Read summary statistics for ieu</code></pre>
</div>
</div>
<div class="cell" data-execution_count="4">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>trait_df <span class="op">=</span> pd.read_csv(trait_df_filename)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>phenotype_dict <span class="op">=</span> trait_df.set_index(<span class="st">'ID'</span>)[<span class="st">'Broad'</span>].to_dict()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="clean" class="level1">
<h1>Clean</h1>
<section id="combine-all-summary-statistics-to-a-single-dataframe." class="level3">
<h3 class="anchored" data-anchor-id="combine-all-summary-statistics-to-a-single-dataframe.">Combine all summary statistics to a single <code>DataFrame</code>.</h3>
<div class="cell" data-execution_count="5">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>assoc_df <span class="op">=</span> pd.concat([v <span class="cf">for</span> k,v <span class="kw">in</span> assoc.items()])</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>assoc_df[<span class="st">'A1'</span>] <span class="op">=</span> assoc_df[<span class="st">'A1'</span>].<span class="bu">str</span>.upper()</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>assoc_df[<span class="st">'A2'</span>] <span class="op">=</span> assoc_df[<span class="st">'A2'</span>].<span class="bu">str</span>.upper()</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>assoc_df[<span class="st">'TRAIT'</span>] <span class="op">=</span> assoc_df[<span class="st">'TRAIT'</span>].fillna(assoc_df[<span class="st">'ID'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-execution_count="6">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>assoc_df <span class="op">=</span> assoc_df.drop_duplicates()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>assoc_df.info()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;class 'pandas.core.frame.DataFrame'&gt;
Index: 680512 entries, 0 to 258531
Data columns (total 11 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   SNP     680512 non-null  string 
 1   CHR     664837 non-null  string 
 2   BP      664837 non-null  Int64  
 3   A2      680512 non-null  string 
 4   A1      680512 non-null  string 
 5   Z       260187 non-null  float64
 6   P       680512 non-null  float64
 7   BETA    640083 non-null  float64
 8   SE      640068 non-null  float64
 9   ID      680512 non-null  string 
 10  TRAIT   680512 non-null  string 
dtypes: Int64(1), float64(4), string(6)
memory usage: 63.0 MB</code></pre>
</div>
</div>
<div class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>assoc_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="8">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">SNP</th>
<th data-quarto-table-cell-role="th">CHR</th>
<th data-quarto-table-cell-role="th">BP</th>
<th data-quarto-table-cell-role="th">A2</th>
<th data-quarto-table-cell-role="th">A1</th>
<th data-quarto-table-cell-role="th">Z</th>
<th data-quarto-table-cell-role="th">P</th>
<th data-quarto-table-cell-role="th">BETA</th>
<th data-quarto-table-cell-role="th">SE</th>
<th data-quarto-table-cell-role="th">ID</th>
<th data-quarto-table-cell-role="th">TRAIT</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>rs147538909</td>
<td>chr1</td>
<td>115746</td>
<td>T</td>
<td>C</td>
<td>-1.197695</td>
<td>0.231036</td>
<td>NaN</td>
<td>NaN</td>
<td>UKB_1160_Sleep_duration</td>
<td>UKB_1160_Sleep_duration</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>rs2977608</td>
<td>chr1</td>
<td>832873</td>
<td>C</td>
<td>A</td>
<td>0.205174</td>
<td>0.837436</td>
<td>0.000454</td>
<td>0.002213</td>
<td>UKB_1160_Sleep_duration</td>
<td>UKB_1160_Sleep_duration</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>rs9442391</td>
<td>chr1</td>
<td>1048922</td>
<td>C</td>
<td>T</td>
<td>1.415529</td>
<td>0.156913</td>
<td>0.002705</td>
<td>0.001911</td>
<td>UKB_1160_Sleep_duration</td>
<td>UKB_1160_Sleep_duration</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>rs6685064</td>
<td>chr1</td>
<td>1275912</td>
<td>T</td>
<td>C</td>
<td>-1.869576</td>
<td>0.061543</td>
<td>-0.007038</td>
<td>0.003764</td>
<td>UKB_1160_Sleep_duration</td>
<td>UKB_1160_Sleep_duration</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>rs79113395</td>
<td>chr1</td>
<td>1659060</td>
<td>A</td>
<td>G</td>
<td>-0.462598</td>
<td>0.643653</td>
<td>-0.000987</td>
<td>0.002134</td>
<td>UKB_1160_Sleep_duration</td>
<td>UKB_1160_Sleep_duration</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">258527</td>
<td>rs7599488</td>
<td>2</td>
<td>60718347</td>
<td>T</td>
<td>C</td>
<td>NaN</td>
<td>0.730000</td>
<td>0.000530</td>
<td>0.001563</td>
<td>ieu-b-13</td>
<td>childhood absence epilepsy</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">258528</td>
<td>rs75010292</td>
<td>2</td>
<td>55172536</td>
<td>G</td>
<td>A</td>
<td>NaN</td>
<td>0.091000</td>
<td>-0.004222</td>
<td>0.002496</td>
<td>ieu-b-13</td>
<td>childhood absence epilepsy</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">258529</td>
<td>rs10168513</td>
<td>2</td>
<td>55452375</td>
<td>A</td>
<td>G</td>
<td>NaN</td>
<td>0.860000</td>
<td>-0.000413</td>
<td>0.002291</td>
<td>ieu-b-13</td>
<td>childhood absence epilepsy</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">258530</td>
<td>rs17428810</td>
<td>2</td>
<td>32736043</td>
<td>C</td>
<td>T</td>
<td>NaN</td>
<td>0.840000</td>
<td>-0.000342</td>
<td>0.001684</td>
<td>ieu-b-13</td>
<td>childhood absence epilepsy</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">258531</td>
<td>rs615449</td>
<td>2</td>
<td>40469480</td>
<td>A</td>
<td>C</td>
<td>NaN</td>
<td>0.420000</td>
<td>-0.002027</td>
<td>0.002493</td>
<td>ieu-b-13</td>
<td>childhood absence epilepsy</td>
</tr>
</tbody>
</table>

<p>680512 rows × 11 columns</p>
</div>
</div>
</div>
<div class="cell" data-execution_count="9">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Number of unique SNPs: </span><span class="sc">{</span><span class="bu">len</span>(assoc_df[<span class="st">'SNP'</span>].unique())<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Number of unique studies: </span><span class="sc">{</span><span class="bu">len</span>(assoc_df[<span class="st">'ID'</span>].unique())<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Number of unique SNPs: 10262
Number of unique studies: 92</code></pre>
</div>
</div>
</section>
<section id="keep-only-biallelic-snps" class="level3">
<h3 class="anchored" data-anchor-id="keep-only-biallelic-snps">Keep only Biallelic SNPs</h3>
<p><code>assoc_df_fa</code>: fa or ‘filter_alleles’ removes alleles which consists of more than two nucleotides.</p>
<div class="cell" data-execution_count="10">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> count_nucleotides(row, df):</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">'''</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co">    df must have 3 columns: SNP, A1, A2</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co">    '''</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    snp <span class="op">=</span> row[<span class="st">'SNP'</span>]</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    snp_dict <span class="op">=</span> df[df[<span class="st">'SNP'</span>] <span class="op">==</span> snp][[<span class="st">'A1'</span>, <span class="st">'A2'</span>]].to_dict(<span class="st">'records'</span>)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> <span class="bu">len</span>(<span class="bu">set</span>([v <span class="cf">for</span> x <span class="kw">in</span> snp_dict <span class="cf">for</span> k, v <span class="kw">in</span> x.items()]))</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="co">'''</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="co">Whether to count the alleles.</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a><span class="co">If a SNP appears twice, most likely it is biallelic with (A1, A2) and (A2, A1) in the two rows.</span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="co">However, if the second row has (A2, A3), then it is not biallelic.</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a><span class="co">Therefore, we count the unique nucleotides for each SNP.</span></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a><span class="co">!!! This is slow. I have checked there are no (A2, A3) for the current data. Hence, avoid.</span></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a><span class="co">'''</span></span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>do_count_nucleotides <span class="op">=</span> <span class="va">False</span></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>alleles       <span class="op">=</span> assoc_df[[<span class="st">'SNP'</span>, <span class="st">'A1'</span>, <span class="st">'A2'</span>]].drop_duplicates()</span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>count_colname <span class="op">=</span> <span class="st">'count'</span></span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>alleles_count <span class="op">=</span> alleles[<span class="st">'SNP'</span>].value_counts().reset_index(name<span class="op">=</span>count_colname)</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> do_count_nucleotides:</span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a>    count_colname <span class="op">=</span> <span class="st">'nucleotide_count'</span></span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>    alleles_count[count_colname] <span class="op">=</span> alleles_count.<span class="bu">apply</span>(count_nucleotides, axis <span class="op">=</span> <span class="dv">1</span>, args <span class="op">=</span> (alleles,))</span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a>biallelic <span class="op">=</span> alleles_count[alleles_count[count_colname] <span class="op">&lt;=</span> <span class="dv">2</span>].merge(alleles, on<span class="op">=</span><span class="st">'SNP'</span>, how<span class="op">=</span><span class="st">'inner'</span>)</span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Arbitrary choice of alleles</span></span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a>one_arrange <span class="op">=</span> biallelic.drop_duplicates(<span class="st">'SNP'</span>, keep<span class="op">=</span><span class="st">'first'</span>).<span class="bu">filter</span>(regex <span class="op">=</span> <span class="st">'^(?!.*count)'</span>, axis <span class="op">=</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-execution_count="11">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>assoc_df_allele1 <span class="op">=</span> one_arrange.merge(assoc_df, on<span class="op">=</span>[<span class="st">'SNP'</span>, <span class="st">'A1'</span>, <span class="st">'A2'</span>], how<span class="op">=</span><span class="st">'inner'</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>assoc_df_allele2 <span class="op">=</span> one_arrange.merge(assoc_df, </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>                        left_on<span class="op">=</span>[<span class="st">'SNP'</span>,  <span class="st">'A1'</span>, <span class="st">'A2'</span>], </span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>                        right_on<span class="op">=</span>[<span class="st">'SNP'</span>, <span class="st">'A2'</span>, <span class="st">'A1'</span>],</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>                        suffixes<span class="op">=</span>(<span class="st">''</span>, <span class="st">'_DROP'</span>)).<span class="bu">filter</span>(regex<span class="op">=</span><span class="st">'^(?!.*_DROP)'</span>)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>assoc_df_allele2[<span class="st">'BETA'</span>] <span class="op">=</span> assoc_df_allele2[<span class="st">'BETA'</span>] <span class="op">*</span> <span class="op">-</span><span class="dv">1</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>assoc_df_allele2[<span class="st">'Z'</span>]    <span class="op">=</span> assoc_df_allele2[<span class="st">'Z'</span>] <span class="op">*</span> <span class="op">-</span><span class="dv">1</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>assoc_df_fa <span class="op">=</span> pd.concat([assoc_df_allele1, assoc_df_allele2]) <span class="co"># fa: fixed alleles</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>assoc_df_fa.info()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;class 'pandas.core.frame.DataFrame'&gt;
Index: 677779 entries, 0 to 105160
Data columns (total 11 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   SNP     677779 non-null  string 
 1   A1      677779 non-null  string 
 2   A2      677779 non-null  string 
 3   CHR     662150 non-null  string 
 4   BP      662150 non-null  Int64  
 5   Z       259246 non-null  float64
 6   P       677779 non-null  float64
 7   BETA    637545 non-null  float64
 8   SE      637530 non-null  float64
 9   ID      677779 non-null  string 
 10  TRAIT   677779 non-null  string 
dtypes: Int64(1), float64(4), string(6)
memory usage: 62.7 MB</code></pre>
</div>
</div>
<div class="cell" data-execution_count="13">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Number of unique SNPs: </span><span class="sc">{</span><span class="bu">len</span>(assoc_df_fa[<span class="st">'SNP'</span>].unique())<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Number of unique studies: </span><span class="sc">{</span><span class="bu">len</span>(assoc_df_fa[<span class="st">'ID'</span>].unique())<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Number of unique SNPs: 10224
Number of unique studies: 92</code></pre>
</div>
</div>
</section>
<section id="remove-snps-found-in-fewer-studies" class="level3">
<h3 class="anchored" data-anchor-id="remove-snps-found-in-fewer-studies">Remove SNPs found in fewer studies</h3>
<p>SNPs which are available in only a few studies are less powerful ‘features’. For a distribution of SNPs, see <a href="#fig-hist-snp-distribution">Figure&nbsp;1</a>.</p>
<div class="cell" data-execution_count="14">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>snps_count <span class="op">=</span> assoc_df_fa[<span class="st">'SNP'</span>].value_counts().reset_index(name<span class="op">=</span><span class="st">'count'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-execution_count="15">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure()</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>ax1 <span class="op">=</span> fig.add_subplot(<span class="dv">111</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>ax1.hist(snps_count[<span class="st">'count'</span>], density <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>ax1.set_xlabel(<span class="st">'Number of studies available for each SNP'</span>)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>ax1.set_ylabel(<span class="st">'Density'</span>)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div id="fig-hist-snp-distribution" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="2023-05-12-preprocess_files/figure-html/fig-hist-snp-distribution-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;1: Distribution of SNPs in different studies</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell" data-execution_count="16">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>assoc_df_fa_nsnp <span class="op">=</span> snps_count[snps_count[<span class="st">'count'</span>] <span class="op">&gt;=</span><span class="dv">20</span> ].merge(assoc_df_fa, on<span class="op">=</span>[<span class="st">'SNP'</span>], how<span class="op">=</span><span class="st">'inner'</span>).<span class="bu">filter</span>(regex <span class="op">=</span> <span class="st">'^(?!.*count)'</span>, axis <span class="op">=</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>assoc_df_fa_nsnp.info()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;class 'pandas.core.frame.DataFrame'&gt;
RangeIndex: 676261 entries, 0 to 676260
Data columns (total 11 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   SNP     676261 non-null  string 
 1   A1      676261 non-null  string 
 2   A2      676261 non-null  string 
 3   CHR     660634 non-null  string 
 4   BP      660634 non-null  Int64  
 5   Z       259054 non-null  float64
 6   P       676261 non-null  float64
 7   BETA    636027 non-null  float64
 8   SE      636012 non-null  float64
 9   ID      676261 non-null  string 
 10  TRAIT   676261 non-null  string 
dtypes: Int64(1), float64(4), string(6)
memory usage: 57.4 MB</code></pre>
</div>
</div>
<div class="cell" data-execution_count="18">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Number of unique SNPs: </span><span class="sc">{</span><span class="bu">len</span>(assoc_df_fa_nsnp[<span class="st">'SNP'</span>].unique())<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Number of unique studies: </span><span class="sc">{</span><span class="bu">len</span>(assoc_df_fa_nsnp[<span class="st">'ID'</span>].unique())<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Number of unique SNPs: 10068
Number of unique studies: 92</code></pre>
</div>
</div>
</section>
<section id="remove-duplicate-snp-id-combinations" class="level3">
<h3 class="anchored" data-anchor-id="remove-duplicate-snp-id-combinations">Remove duplicate (SNP, ID) combinations</h3>
<div class="cell" data-execution_count="19">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>assoc_df_fa_nsnp_nodup <span class="op">=</span> assoc_df_fa_nsnp.drop_duplicates(subset<span class="op">=</span>[<span class="st">'SNP'</span>, <span class="st">'ID'</span>], keep <span class="op">=</span> <span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>assoc_df_fa_nsnp_nodup.info()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;class 'pandas.core.frame.DataFrame'&gt;
Index: 676229 entries, 0 to 676260
Data columns (total 11 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   SNP     676229 non-null  string 
 1   A1      676229 non-null  string 
 2   A2      676229 non-null  string 
 3   CHR     660602 non-null  string 
 4   BP      660602 non-null  Int64  
 5   Z       259054 non-null  float64
 6   P       676229 non-null  float64
 7   BETA    635995 non-null  float64
 8   SE      635980 non-null  float64
 9   ID      676229 non-null  string 
 10  TRAIT   676229 non-null  string 
dtypes: Int64(1), float64(4), string(6)
memory usage: 62.6 MB</code></pre>
</div>
</div>
<div class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>assoc_df_fa_nsnp_nodup</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="21">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">SNP</th>
<th data-quarto-table-cell-role="th">A1</th>
<th data-quarto-table-cell-role="th">A2</th>
<th data-quarto-table-cell-role="th">CHR</th>
<th data-quarto-table-cell-role="th">BP</th>
<th data-quarto-table-cell-role="th">Z</th>
<th data-quarto-table-cell-role="th">P</th>
<th data-quarto-table-cell-role="th">BETA</th>
<th data-quarto-table-cell-role="th">SE</th>
<th data-quarto-table-cell-role="th">ID</th>
<th data-quarto-table-cell-role="th">TRAIT</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>rs10486722</td>
<td>C</td>
<td>T</td>
<td>chr7</td>
<td>41772310</td>
<td>0.881222</td>
<td>0.378198</td>
<td>0.001757</td>
<td>0.001994</td>
<td>UKB_1160_Sleep_duration</td>
<td>UKB_1160_Sleep_duration</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>rs10486722</td>
<td>C</td>
<td>T</td>
<td>chr7</td>
<td>41772310</td>
<td>1.369354</td>
<td>0.170889</td>
<td>0.003476</td>
<td>0.002538</td>
<td>UKB_1180_Morning_or_evening_person_chronotype</td>
<td>UKB_1180_Morning_or_evening_person_chronotype</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>rs10486722</td>
<td>C</td>
<td>T</td>
<td>chr7</td>
<td>41772310</td>
<td>-0.138782</td>
<td>0.889622</td>
<td>-0.000257</td>
<td>0.001850</td>
<td>UKB_1200_Sleeplessness_or_insomnia</td>
<td>UKB_1200_Sleeplessness_or_insomnia</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>rs10486722</td>
<td>C</td>
<td>T</td>
<td>chr7</td>
<td>41772310</td>
<td>0.442336</td>
<td>0.658246</td>
<td>0.000035</td>
<td>0.000080</td>
<td>UKB_20002_1243_self_reported_psychological_or_...</td>
<td>UKB_20002_1243_self_reported_psychological_or_...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>rs10486722</td>
<td>C</td>
<td>T</td>
<td>chr7</td>
<td>41772310</td>
<td>0.275442</td>
<td>0.782977</td>
<td>0.000030</td>
<td>0.000110</td>
<td>UKB_20002_1262_self_reported_parkinsons_disease</td>
<td>UKB_20002_1262_self_reported_parkinsons_disease</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">676256</td>
<td>rs2322</td>
<td>C</td>
<td>G</td>
<td>4</td>
<td>9701603</td>
<td>NaN</td>
<td>0.396100</td>
<td>-0.077600</td>
<td>0.091500</td>
<td>ieu-b-7</td>
<td>Parkinson's disease</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">676257</td>
<td>rs2322</td>
<td>C</td>
<td>G</td>
<td>4</td>
<td>9701603</td>
<td>NaN</td>
<td>0.107599</td>
<td>0.024041</td>
<td>0.014941</td>
<td>MHQ_Depression_WG_MAF1_INFO4_HRC_Only_Filtered...</td>
<td>MHQ_Depression_WG_MAF1_INFO4_HRC_Only_Filtered...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">676258</td>
<td>rs2322</td>
<td>C</td>
<td>G</td>
<td>4</td>
<td>9701603</td>
<td>NaN</td>
<td>0.430338</td>
<td>0.016141</td>
<td>0.020468</td>
<td>MHQ_Recurrent_Depression_WG_MAF1_INFO4_HRC_Onl...</td>
<td>MHQ_Recurrent_Depression_WG_MAF1_INFO4_HRC_Onl...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">676259</td>
<td>rs2322</td>
<td>C</td>
<td>G</td>
<td>4</td>
<td>9701603</td>
<td>NaN</td>
<td>0.068770</td>
<td>0.049301</td>
<td>0.027090</td>
<td>MHQ_Single_Depression_WG_MAF1_INFO4_HRC_Only_F...</td>
<td>MHQ_Single_Depression_WG_MAF1_INFO4_HRC_Only_F...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">676260</td>
<td>rs2322</td>
<td>C</td>
<td>G</td>
<td>4</td>
<td>9701603</td>
<td>NaN</td>
<td>0.721340</td>
<td>0.006329</td>
<td>0.017746</td>
<td>MHQ_Subthreshold_WG_MAF1_INFO4_HRC_Only_Filter...</td>
<td>MHQ_Subthreshold_WG_MAF1_INFO4_HRC_Only_Filter...</td>
</tr>
</tbody>
</table>

<p>676229 rows × 11 columns</p>
</div>
</div>
</div>
<div class="cell" data-execution_count="22">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Number of unique SNPs: </span><span class="sc">{</span><span class="bu">len</span>(assoc_df_fa_nsnp_nodup[<span class="st">'SNP'</span>].unique())<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Number of unique studies: </span><span class="sc">{</span><span class="bu">len</span>(assoc_df_fa_nsnp_nodup[<span class="st">'ID'</span>].unique())<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Number of unique SNPs: 10068
Number of unique studies: 92</code></pre>
</div>
</div>
</section>
</section>
<section id="data-matrix" class="level1">
<h1>Data Matrix</h1>
<div class="cell" data-execution_count="23">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>beta_df   <span class="op">=</span> assoc_df_fa_nsnp_nodup[[<span class="st">'SNP'</span>, <span class="st">'ID'</span>, <span class="st">'BETA'</span>]].pivot(index <span class="op">=</span> <span class="st">'SNP'</span>, columns <span class="op">=</span> <span class="st">'ID'</span>, values <span class="op">=</span> <span class="st">'BETA'</span>).rename_axis(<span class="va">None</span>, axis <span class="op">=</span> <span class="dv">0</span>).rename_axis(<span class="va">None</span>, axis <span class="op">=</span> <span class="dv">1</span>)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>se_df     <span class="op">=</span> assoc_df_fa_nsnp_nodup[[<span class="st">'SNP'</span>, <span class="st">'ID'</span>, <span class="st">'SE'</span>]].pivot(index <span class="op">=</span> <span class="st">'SNP'</span>, columns <span class="op">=</span> <span class="st">'ID'</span>, values <span class="op">=</span> <span class="st">'SE'</span>).rename_axis(<span class="va">None</span>, axis <span class="op">=</span> <span class="dv">0</span>).rename_axis(<span class="va">None</span>, axis <span class="op">=</span> <span class="dv">1</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>zscore_df <span class="op">=</span> assoc_df_fa_nsnp_nodup[[<span class="st">'SNP'</span>, <span class="st">'ID'</span>, <span class="st">'Z'</span>]].pivot(index <span class="op">=</span> <span class="st">'SNP'</span>, columns <span class="op">=</span> <span class="st">'ID'</span>, values <span class="op">=</span> <span class="st">'Z'</span>).rename_axis(<span class="va">None</span>, axis <span class="op">=</span> <span class="dv">0</span>).rename_axis(<span class="va">None</span>, axis <span class="op">=</span> <span class="dv">1</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>se_df     <span class="op">=</span> se_df.fillna(beta_df <span class="op">/</span> zscore_df).replace(<span class="dv">0</span>, np.nan)</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>zscore_df <span class="op">=</span> zscore_df.fillna(beta_df <span class="op">/</span> se_df)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>prec_df   <span class="op">=</span> se_df.<span class="bu">apply</span>(<span class="kw">lambda</span> x: <span class="dv">1</span> <span class="op">/</span> x <span class="op">/</span> x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>zscore_df.info()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;class 'pandas.core.frame.DataFrame'&gt;
Index: 10068 entries, rs1000031 to rs999494
Data columns (total 92 columns):
 #   Column                                                                                            Non-Null Count  Dtype  
---  ------                                                                                            --------------  -----  
 0   AD_sumstats_Jansenetal_2019sept.txt.gz                                                            10020 non-null  float64
 1   CNCR_Insomnia_all                                                                                 9792 non-null   float64
 2   ENIGMA_Intracraneal_Volume                                                                        9706 non-null   float64
 3   GPC-NEO-NEUROTICISM                                                                               9430 non-null   float64
 4   IGAP_Alzheimer                                                                                    9651 non-null   float64
 5   ILAE_Genetic_generalised_epilepsy                                                                 9484 non-null   float64
 6   Jones_et_al_2016_Chronotype                                                                       9782 non-null   float64
 7   Jones_et_al_2016_SleepDuration                                                                    9782 non-null   float64
 8   MDD_MHQ_BIP_METACARPA_INFO6_A5_NTOT_no23andMe_noUKBB.txt.gz                                       7801 non-null   float64
 9   MDD_MHQ_METACARPA_INFO6_A5_NTOT_no23andMe_noUKBB.txt.gz                                           7826 non-null   float64
 10  MHQ_Depression_WG_MAF1_INFO4_HRC_Only_Filtered_Dups_FOR_METACARPA_INFO6_A5_NTOT.txt.gz            7977 non-null   float64
 11  MHQ_Recurrent_Depression_WG_MAF1_INFO4_HRC_Only_Filtered_Dups_FOR_METACARPA_INFO6_A5_NTOT.txt.gz  7971 non-null   float64
 12  MHQ_Single_Depression_WG_MAF1_INFO4_HRC_Only_Filtered_Dups_FOR_METACARPA_INFO6_A5_NTOT.txt.gz     7972 non-null   float64
 13  MHQ_Subthreshold_WG_MAF1_INFO4_HRC_Only_Filtered_Dups_FOR_METACARPA_INFO6_A5_NTOT.txt.gz          7973 non-null   float64
 14  PGC3_SCZ_wave3_public.v2.txt.gz                                                                   9707 non-null   float64
 15  PGC_ADHD_EUR_2017                                                                                 9688 non-null   float64
 16  PGC_ASD_2017_CEU                                                                                  9626 non-null   float64
 17  SSGAC_Depressive_Symptoms                                                                         9672 non-null   float64
 18  SSGAC_Education_Years_Pooled                                                                      9740 non-null   float64
 19  UKB_1160_Sleep_duration                                                                           9776 non-null   float64
 20  UKB_1180_Morning_or_evening_person_chronotype                                                     9776 non-null   float64
 21  UKB_1200_Sleeplessness_or_insomnia                                                                9776 non-null   float64
 22  UKB_20002_1243_self_reported_psychological_or_psychiatric_problem                                 9776 non-null   float64
 23  UKB_20002_1262_self_reported_parkinsons_disease                                                   9776 non-null   float64
 24  UKB_20002_1265_self_reported_migraine                                                             9776 non-null   float64
 25  UKB_20002_1289_self_reported_schizophrenia                                                        9776 non-null   float64
 26  UKB_20002_1616_self_reported_insomnia                                                             9776 non-null   float64
 27  UKB_20016_Fluid_intelligence_score                                                                9776 non-null   float64
 28  UKB_20127_Neuroticism_score                                                                       9776 non-null   float64
 29  UKB_G40_Diagnoses_main_ICD10_G40_Epilepsy                                                         9776 non-null   float64
 30  UKB_G43_Diagnoses_main_ICD10_G43_Migraine                                                         9776 non-null   float64
 31  anxiety.meta.full.cc.txt.gz                                                                       8449 non-null   float64
 32  anxiety.meta.full.fs.txt.gz                                                                       8418 non-null   float64
 33  daner_PGC_BIP32b_mds7a_0416a.txt.gz                                                               9988 non-null   float64
 34  daner_PGC_BIP32b_mds7a_mds7a_BD1.0416a_INFO6_A5_NTOT.txt.gz                                       8108 non-null   float64
 35  daner_PGC_BIP32b_mds7a_mds7a_BD2.0416a_INFO6_A5_NTOT.txt.gz                                       8096 non-null   float64
 36  daner_adhd_meta_filtered_NA_iPSYCH23_PGC11_sigPCs_woSEX_2ell6sd_EUR_Neff_70.txt.gz                8795 non-null   float64
 37  iPSYCH-PGC_ASD_Nov2017.txt.gz                                                                     9654 non-null   float64
 38  ieu-a-1000                                                                                        8792 non-null   float64
 39  ieu-a-1009                                                                                        4022 non-null   float64
 40  ieu-a-1018                                                                                        4009 non-null   float64
 41  ieu-a-1019                                                                                        1439 non-null   float64
 42  ieu-a-1029                                                                                        4180 non-null   float64
 43  ieu-a-1041                                                                                        8494 non-null   float64
 44  ieu-a-1042                                                                                        8492 non-null   float64
 45  ieu-a-1043                                                                                        8496 non-null   float64
 46  ieu-a-1044                                                                                        8495 non-null   float64
 47  ieu-a-1045                                                                                        8496 non-null   float64
 48  ieu-a-1046                                                                                        8491 non-null   float64
 49  ieu-a-1047                                                                                        8496 non-null   float64
 50  ieu-a-1048                                                                                        8498 non-null   float64
 51  ieu-a-1061                                                                                        4117 non-null   float64
 52  ieu-a-1062                                                                                        3940 non-null   float64
 53  ieu-a-1063                                                                                        4059 non-null   float64
 54  ieu-a-1064                                                                                        4120 non-null   float64
 55  ieu-a-1065                                                                                        4117 non-null   float64
 56  ieu-a-1066                                                                                        4120 non-null   float64
 57  ieu-a-1067                                                                                        4116 non-null   float64
 58  ieu-a-1068                                                                                        4117 non-null   float64
 59  ieu-a-1085                                                                                        9043 non-null   float64
 60  ieu-a-118                                                                                         8939 non-null   float64
 61  ieu-a-297                                                                                         8870 non-null   float64
 62  ieu-a-298                                                                                         242 non-null    float64
 63  ieu-a-45                                                                                          2539 non-null   float64
 64  ieu-a-808                                                                                         789 non-null    float64
 65  ieu-a-810                                                                                         702 non-null    float64
 66  ieu-a-812                                                                                         1206 non-null   float64
 67  ieu-a-818                                                                                         377 non-null    float64
 68  ieu-a-824                                                                                         719 non-null    float64
 69  ieu-a-990                                                                                         7676 non-null   float64
 70  ieu-b-10                                                                                          5586 non-null   float64
 71  ieu-b-11                                                                                          5817 non-null   float64
 72  ieu-b-12                                                                                          5828 non-null   float64
 73  ieu-b-13                                                                                          5820 non-null   float64
 74  ieu-b-14                                                                                          5780 non-null   float64
 75  ieu-b-15                                                                                          5821 non-null   float64
 76  ieu-b-16                                                                                          5829 non-null   float64
 77  ieu-b-17                                                                                          5814 non-null   float64
 78  ieu-b-18                                                                                          7781 non-null   float64
 79  ieu-b-2                                                                                           9360 non-null   float64
 80  ieu-b-43                                                                                          964 non-null    float64
 81  ieu-b-5070                                                                                        7585 non-null   float64
 82  ieu-b-7                                                                                           9227 non-null   float64
 83  ieu-b-8                                                                                           5508 non-null   float64
 84  ieu-b-9                                                                                           5631 non-null   float64
 85  ocd_aug2017.txt.gz                                                                                9774 non-null   float64
 86  pgc-bip2021-BDI.vcf.txt.gz                                                                        9551 non-null   float64
 87  pgc-bip2021-BDII.vcf.txt.gz                                                                       9418 non-null   float64
 88  pgc-bip2021-all.vcf.txt.gz                                                                        9718 non-null   float64
 89  pgc.scz2                                                                                          9742 non-null   float64
 90  pgcAN2.2019-07.vcf.txt.gz                                                                         9057 non-null   float64
 91  pts_all_freeze2_overall.txt.gz                                                                    9988 non-null   float64
dtypes: float64(92)
memory usage: 7.1 MB</code></pre>
</div>
</div>
<div class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>pd.testing.assert_index_equal(beta_df.index, prec_df.index)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>pd.testing.assert_index_equal(beta_df.index, zscore_df.index)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>pd.testing.assert_index_equal(beta_df.columns, prec_df.columns)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>pd.testing.assert_index_equal(beta_df.columns, zscore_df.columns)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="remove-studies-with-missingness" class="level3">
<h3 class="anchored" data-anchor-id="remove-studies-with-missingness">Remove studies with missingness</h3>
<p>Studies with high missingness are not good samples, hence they are removed. See histogram of missingness in <a href="#fig-missingness">Figure&nbsp;2</a></p>
<div class="cell" data-execution_count="26">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>zscore_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="26">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">AD_sumstats_Jansenetal_2019sept.txt.gz</th>
<th data-quarto-table-cell-role="th">CNCR_Insomnia_all</th>
<th data-quarto-table-cell-role="th">ENIGMA_Intracraneal_Volume</th>
<th data-quarto-table-cell-role="th">GPC-NEO-NEUROTICISM</th>
<th data-quarto-table-cell-role="th">IGAP_Alzheimer</th>
<th data-quarto-table-cell-role="th">ILAE_Genetic_generalised_epilepsy</th>
<th data-quarto-table-cell-role="th">Jones_et_al_2016_Chronotype</th>
<th data-quarto-table-cell-role="th">Jones_et_al_2016_SleepDuration</th>
<th data-quarto-table-cell-role="th">MDD_MHQ_BIP_METACARPA_INFO6_A5_NTOT_no23andMe_...</th>
<th data-quarto-table-cell-role="th">MDD_MHQ_METACARPA_INFO6_A5_NTOT_no23andMe_noUK...</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">ieu-b-7</th>
<th data-quarto-table-cell-role="th">ieu-b-8</th>
<th data-quarto-table-cell-role="th">ieu-b-9</th>
<th data-quarto-table-cell-role="th">ocd_aug2017.txt.gz</th>
<th data-quarto-table-cell-role="th">pgc-bip2021-BDI.vcf.txt.gz</th>
<th data-quarto-table-cell-role="th">pgc-bip2021-BDII.vcf.txt.gz</th>
<th data-quarto-table-cell-role="th">pgc-bip2021-all.vcf.txt.gz</th>
<th data-quarto-table-cell-role="th">pgc.scz2</th>
<th data-quarto-table-cell-role="th">pgcAN2.2019-07.vcf.txt.gz</th>
<th data-quarto-table-cell-role="th">pts_all_freeze2_overall.txt.gz</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">rs1000031</td>
<td>0.999531</td>
<td>-0.327477</td>
<td>2.184712</td>
<td>1.241557</td>
<td>0.441709</td>
<td>-1.041816</td>
<td>0.163658</td>
<td>-0.163658</td>
<td>0.336654</td>
<td>0.793129</td>
<td>...</td>
<td>0.532189</td>
<td>NaN</td>
<td>NaN</td>
<td>-0.198735</td>
<td>1.057089</td>
<td>-0.269020</td>
<td>1.279776</td>
<td>-0.433158</td>
<td>-1.573766</td>
<td>-1.674269</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">rs1000269</td>
<td>1.212805</td>
<td>-1.046310</td>
<td>0.001880</td>
<td>0.741814</td>
<td>-1.844296</td>
<td>0.771000</td>
<td>2.673787</td>
<td>1.126391</td>
<td>-0.092067</td>
<td>-0.163246</td>
<td>...</td>
<td>1.665179</td>
<td>-0.732000</td>
<td>-0.699000</td>
<td>0.100883</td>
<td>-0.226381</td>
<td>0.338368</td>
<td>-0.924392</td>
<td>0.832016</td>
<td>0.681645</td>
<td>-0.701776</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">rs10003281</td>
<td>0.813444</td>
<td>2.034345</td>
<td>-2.023031</td>
<td>-1.750164</td>
<td>-0.076778</td>
<td>1.448634</td>
<td>0.954165</td>
<td>-1.805477</td>
<td>NaN</td>
<td>NaN</td>
<td>...</td>
<td>-0.475795</td>
<td>4.437998</td>
<td>2.366001</td>
<td>0.967399</td>
<td>0.286699</td>
<td>-1.162661</td>
<td>-0.199299</td>
<td>0.014539</td>
<td>NaN</td>
<td>-1.379710</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">rs10004866</td>
<td>-0.011252</td>
<td>1.327108</td>
<td>1.004786</td>
<td>1.442363</td>
<td>-1.215173</td>
<td>0.139000</td>
<td>0.050154</td>
<td>1.439531</td>
<td>-2.458370</td>
<td>-2.407460</td>
<td>...</td>
<td>-1.234375</td>
<td>-2.520001</td>
<td>-0.593997</td>
<td>-0.685110</td>
<td>0.902252</td>
<td>1.106939</td>
<td>1.776456</td>
<td>-1.654677</td>
<td>-0.964630</td>
<td>0.851608</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">rs10005235</td>
<td>-0.612540</td>
<td>-0.410609</td>
<td>1.526040</td>
<td>0.653087</td>
<td>0.344062</td>
<td>-1.868000</td>
<td>2.183486</td>
<td>-1.514102</td>
<td>0.460191</td>
<td>0.393006</td>
<td>...</td>
<td>0.387805</td>
<td>-0.345000</td>
<td>-0.960998</td>
<td>0.177317</td>
<td>-1.339598</td>
<td>1.795867</td>
<td>-1.249969</td>
<td>2.349671</td>
<td>0.996305</td>
<td>-0.333356</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">rs9989571</td>
<td>-0.028306</td>
<td>-0.208891</td>
<td>1.883150</td>
<td>0.366470</td>
<td>0.821257</td>
<td>0.712000</td>
<td>-0.453762</td>
<td>1.895698</td>
<td>-0.218149</td>
<td>-0.920789</td>
<td>...</td>
<td>-1.231511</td>
<td>-0.820996</td>
<td>0.712000</td>
<td>-2.150176</td>
<td>-0.877410</td>
<td>-1.938969</td>
<td>-2.729983</td>
<td>3.207917</td>
<td>1.469194</td>
<td>-1.293122</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">rs9991694</td>
<td>0.679790</td>
<td>-1.005571</td>
<td>-1.004740</td>
<td>0.753472</td>
<td>-0.539271</td>
<td>0.368971</td>
<td>-1.674665</td>
<td>2.862736</td>
<td>3.744820</td>
<td>3.583060</td>
<td>...</td>
<td>0.064417</td>
<td>NaN</td>
<td>NaN</td>
<td>-2.884911</td>
<td>-1.000231</td>
<td>0.031860</td>
<td>-1.248222</td>
<td>2.309425</td>
<td>NaN</td>
<td>1.048454</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">rs9992763</td>
<td>-0.691405</td>
<td>-0.010299</td>
<td>2.067894</td>
<td>-0.140010</td>
<td>-0.419843</td>
<td>1.320000</td>
<td>0.138304</td>
<td>-0.568052</td>
<td>-0.019684</td>
<td>0.194404</td>
<td>...</td>
<td>0.191860</td>
<td>-0.074000</td>
<td>1.030997</td>
<td>-0.228287</td>
<td>-0.051297</td>
<td>0.781766</td>
<td>0.010638</td>
<td>0.456681</td>
<td>-0.503370</td>
<td>-1.435277</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">rs9993607</td>
<td>1.625392</td>
<td>-0.391585</td>
<td>-0.113291</td>
<td>0.514268</td>
<td>0.027576</td>
<td>-0.512000</td>
<td>-0.150969</td>
<td>0.113039</td>
<td>4.638940</td>
<td>4.631950</td>
<td>...</td>
<td>-0.685106</td>
<td>0.194000</td>
<td>0.240001</td>
<td>-0.790290</td>
<td>-0.876804</td>
<td>-0.577696</td>
<td>-0.785670</td>
<td>-0.062707</td>
<td>0.240834</td>
<td>-0.199740</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">rs999494</td>
<td>0.303642</td>
<td>0.872613</td>
<td>1.522435</td>
<td>-0.227674</td>
<td>1.390424</td>
<td>-1.022395</td>
<td>-0.138304</td>
<td>-1.281552</td>
<td>-1.873050</td>
<td>-1.645590</td>
<td>...</td>
<td>-0.437500</td>
<td>0.253000</td>
<td>-0.926997</td>
<td>-1.449674</td>
<td>0.910515</td>
<td>0.783853</td>
<td>1.376043</td>
<td>-5.195746</td>
<td>-1.151316</td>
<td>0.660120</td>
</tr>
</tbody>
</table>

<p>10068 rows × 92 columns</p>
</div>
</div>
</div>
<div class="cell" data-execution_count="27">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>pd.DataFrame(zscore_df.isnull()).mean(axis<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="27">
<pre><code>AD_sumstats_Jansenetal_2019sept.txt.gz    0.004768
CNCR_Insomnia_all                         0.027414
ENIGMA_Intracraneal_Volume                0.035956
GPC-NEO-NEUROTICISM                       0.063369
IGAP_Alzheimer                            0.041418
                                            ...   
pgc-bip2021-BDII.vcf.txt.gz               0.064561
pgc-bip2021-all.vcf.txt.gz                0.034764
pgc.scz2                                  0.032380
pgcAN2.2019-07.vcf.txt.gz                 0.100417
pts_all_freeze2_overall.txt.gz            0.007946
Length: 92, dtype: float64</code></pre>
</div>
</div>
<div class="cell" data-execution_count="28">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>missing_df <span class="op">=</span> pd.DataFrame(zscore_df.isnull()).mean(axis <span class="op">=</span> <span class="dv">0</span>).reset_index(name <span class="op">=</span> <span class="st">'missingness'</span>).rename(columns <span class="op">=</span> {<span class="st">'index'</span>: <span class="st">'ID'</span>})</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure()</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>ax1 <span class="op">=</span> fig.add_subplot(<span class="dv">111</span>)</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>ax1.hist(missing_df[<span class="st">'missingness'</span>], density <span class="op">=</span> <span class="va">True</span>, bins <span class="op">=</span> <span class="dv">20</span>)</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>ax1.set_xlabel(<span class="st">'Missingness'</span>)</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>ax1.set_ylabel(<span class="st">'Density'</span>)</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div id="fig-missingness" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="2023-05-12-preprocess_files/figure-html/fig-missingness-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;2: Missingness of SNPs in different studies</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell" data-execution_count="29">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>keep_columns <span class="op">=</span> missing_df[missing_df[<span class="st">'missingness'</span>] <span class="op">&lt;=</span> <span class="fl">0.6</span>][<span class="st">'ID'</span>].to_list()</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Keeping </span><span class="sc">{</span><span class="bu">len</span>(keep_columns)<span class="sc">}</span><span class="ss"> phenotypes with low missingness"</span>)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>beta_df_sub   <span class="op">=</span> beta_df[keep_columns]</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>prec_df_sub   <span class="op">=</span> prec_df[keep_columns]</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>zscore_df_sub <span class="op">=</span> zscore_df[keep_columns]</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>se_df_sub     <span class="op">=</span> se_df[keep_columns]</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>remove_columns <span class="op">=</span> [x <span class="cf">for</span> x <span class="kw">in</span> missing_df[<span class="st">'ID'</span>].to_list() <span class="cf">if</span> x <span class="kw">not</span> <span class="kw">in</span> keep_columns]</span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Phenotypes removed:"</span>)</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> x <span class="kw">in</span> remove_columns:</span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>    missingness <span class="op">=</span> missing_df.loc[missing_df[<span class="st">'ID'</span>] <span class="op">==</span> x][<span class="st">'missingness'</span>].iloc[<span class="dv">0</span>]</span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="sc">{</span>missingness<span class="sc">:.3f}</span><span class="ch">\t</span><span class="sc">{</span>x<span class="sc">}</span><span class="ch">\t</span><span class="sc">{</span>phenotype_dict[x]<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Keeping 80 phenotypes with low missingness
Phenotypes removed:
    0.601   ieu-a-1009  Other psych
    0.602   ieu-a-1018  Other psych
    0.857   ieu-a-1019  BD
    0.609   ieu-a-1062  Cognition
    0.976   ieu-a-298   Neurodegenerative
    0.748   ieu-a-45    Other psych
    0.922   ieu-a-808   BD
    0.930   ieu-a-810   SZ
    0.880   ieu-a-812   Neurodegenerative
    0.963   ieu-a-818   Neurodegenerative
    0.929   ieu-a-824   Neurodegenerative
    0.904   ieu-b-43    Other psych</code></pre>
</div>
</div>
<div class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>beta_df_sub.info()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;class 'pandas.core.frame.DataFrame'&gt;
Index: 10068 entries, rs1000031 to rs999494
Data columns (total 80 columns):
 #   Column                                                                                            Non-Null Count  Dtype  
---  ------                                                                                            --------------  -----  
 0   AD_sumstats_Jansenetal_2019sept.txt.gz                                                            10020 non-null  float64
 1   CNCR_Insomnia_all                                                                                 8950 non-null   float64
 2   ENIGMA_Intracraneal_Volume                                                                        8351 non-null   float64
 3   GPC-NEO-NEUROTICISM                                                                               3623 non-null   float64
 4   IGAP_Alzheimer                                                                                    8033 non-null   float64
 5   ILAE_Genetic_generalised_epilepsy                                                                 0 non-null      float64
 6   Jones_et_al_2016_Chronotype                                                                       8807 non-null   float64
 7   Jones_et_al_2016_SleepDuration                                                                    8807 non-null   float64
 8   MDD_MHQ_BIP_METACARPA_INFO6_A5_NTOT_no23andMe_noUKBB.txt.gz                                       7801 non-null   float64
 9   MDD_MHQ_METACARPA_INFO6_A5_NTOT_no23andMe_noUKBB.txt.gz                                           7826 non-null   float64
 10  MHQ_Depression_WG_MAF1_INFO4_HRC_Only_Filtered_Dups_FOR_METACARPA_INFO6_A5_NTOT.txt.gz            7977 non-null   float64
 11  MHQ_Recurrent_Depression_WG_MAF1_INFO4_HRC_Only_Filtered_Dups_FOR_METACARPA_INFO6_A5_NTOT.txt.gz  7971 non-null   float64
 12  MHQ_Single_Depression_WG_MAF1_INFO4_HRC_Only_Filtered_Dups_FOR_METACARPA_INFO6_A5_NTOT.txt.gz     7972 non-null   float64
 13  MHQ_Subthreshold_WG_MAF1_INFO4_HRC_Only_Filtered_Dups_FOR_METACARPA_INFO6_A5_NTOT.txt.gz          7973 non-null   float64
 14  PGC3_SCZ_wave3_public.v2.txt.gz                                                                   9707 non-null   float64
 15  PGC_ADHD_EUR_2017                                                                                 7880 non-null   float64
 16  PGC_ASD_2017_CEU                                                                                  7287 non-null   float64
 17  SSGAC_Depressive_Symptoms                                                                         7888 non-null   float64
 18  SSGAC_Education_Years_Pooled                                                                      8824 non-null   float64
 19  UKB_1160_Sleep_duration                                                                           8829 non-null   float64
 20  UKB_1180_Morning_or_evening_person_chronotype                                                     8829 non-null   float64
 21  UKB_1200_Sleeplessness_or_insomnia                                                                8829 non-null   float64
 22  UKB_20002_1243_self_reported_psychological_or_psychiatric_problem                                 8829 non-null   float64
 23  UKB_20002_1262_self_reported_parkinsons_disease                                                   8829 non-null   float64
 24  UKB_20002_1265_self_reported_migraine                                                             8829 non-null   float64
 25  UKB_20002_1289_self_reported_schizophrenia                                                        8829 non-null   float64
 26  UKB_20002_1616_self_reported_insomnia                                                             8829 non-null   float64
 27  UKB_20016_Fluid_intelligence_score                                                                8829 non-null   float64
 28  UKB_20127_Neuroticism_score                                                                       8829 non-null   float64
 29  UKB_G40_Diagnoses_main_ICD10_G40_Epilepsy                                                         8829 non-null   float64
 30  UKB_G43_Diagnoses_main_ICD10_G43_Migraine                                                         8829 non-null   float64
 31  anxiety.meta.full.cc.txt.gz                                                                       8449 non-null   float64
 32  anxiety.meta.full.fs.txt.gz                                                                       8418 non-null   float64
 33  daner_PGC_BIP32b_mds7a_0416a.txt.gz                                                               9988 non-null   float64
 34  daner_PGC_BIP32b_mds7a_mds7a_BD1.0416a_INFO6_A5_NTOT.txt.gz                                       8108 non-null   float64
 35  daner_PGC_BIP32b_mds7a_mds7a_BD2.0416a_INFO6_A5_NTOT.txt.gz                                       8096 non-null   float64
 36  daner_adhd_meta_filtered_NA_iPSYCH23_PGC11_sigPCs_woSEX_2ell6sd_EUR_Neff_70.txt.gz                8795 non-null   float64
 37  iPSYCH-PGC_ASD_Nov2017.txt.gz                                                                     9654 non-null   float64
 38  ieu-a-1000                                                                                        8792 non-null   float64
 39  ieu-a-1029                                                                                        4180 non-null   float64
 40  ieu-a-1041                                                                                        8494 non-null   float64
 41  ieu-a-1042                                                                                        8492 non-null   float64
 42  ieu-a-1043                                                                                        8496 non-null   float64
 43  ieu-a-1044                                                                                        8495 non-null   float64
 44  ieu-a-1045                                                                                        8496 non-null   float64
 45  ieu-a-1046                                                                                        8491 non-null   float64
 46  ieu-a-1047                                                                                        8496 non-null   float64
 47  ieu-a-1048                                                                                        8498 non-null   float64
 48  ieu-a-1061                                                                                        4117 non-null   float64
 49  ieu-a-1063                                                                                        4059 non-null   float64
 50  ieu-a-1064                                                                                        4120 non-null   float64
 51  ieu-a-1065                                                                                        4117 non-null   float64
 52  ieu-a-1066                                                                                        4120 non-null   float64
 53  ieu-a-1067                                                                                        4116 non-null   float64
 54  ieu-a-1068                                                                                        4117 non-null   float64
 55  ieu-a-1085                                                                                        9043 non-null   float64
 56  ieu-a-118                                                                                         8939 non-null   float64
 57  ieu-a-297                                                                                         8870 non-null   float64
 58  ieu-a-990                                                                                         7676 non-null   float64
 59  ieu-b-10                                                                                          5586 non-null   float64
 60  ieu-b-11                                                                                          5817 non-null   float64
 61  ieu-b-12                                                                                          5828 non-null   float64
 62  ieu-b-13                                                                                          5820 non-null   float64
 63  ieu-b-14                                                                                          5780 non-null   float64
 64  ieu-b-15                                                                                          5821 non-null   float64
 65  ieu-b-16                                                                                          5829 non-null   float64
 66  ieu-b-17                                                                                          5814 non-null   float64
 67  ieu-b-18                                                                                          7781 non-null   float64
 68  ieu-b-2                                                                                           9360 non-null   float64
 69  ieu-b-5070                                                                                        7585 non-null   float64
 70  ieu-b-7                                                                                           9227 non-null   float64
 71  ieu-b-8                                                                                           5508 non-null   float64
 72  ieu-b-9                                                                                           5631 non-null   float64
 73  ocd_aug2017.txt.gz                                                                                9774 non-null   float64
 74  pgc-bip2021-BDI.vcf.txt.gz                                                                        9551 non-null   float64
 75  pgc-bip2021-BDII.vcf.txt.gz                                                                       9418 non-null   float64
 76  pgc-bip2021-all.vcf.txt.gz                                                                        9718 non-null   float64
 77  pgc.scz2                                                                                          8775 non-null   float64
 78  pgcAN2.2019-07.vcf.txt.gz                                                                         9057 non-null   float64
 79  pts_all_freeze2_overall.txt.gz                                                                    9988 non-null   float64
dtypes: float64(80)
memory usage: 6.2 MB</code></pre>
</div>
</div>
<div class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>prec_df_sub.info()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;class 'pandas.core.frame.DataFrame'&gt;
Index: 10068 entries, rs1000031 to rs999494
Data columns (total 80 columns):
 #   Column                                                                                            Non-Null Count  Dtype  
---  ------                                                                                            --------------  -----  
 0   AD_sumstats_Jansenetal_2019sept.txt.gz                                                            10020 non-null  float64
 1   CNCR_Insomnia_all                                                                                 8950 non-null   float64
 2   ENIGMA_Intracraneal_Volume                                                                        8351 non-null   float64
 3   GPC-NEO-NEUROTICISM                                                                               3623 non-null   float64
 4   IGAP_Alzheimer                                                                                    8033 non-null   float64
 5   ILAE_Genetic_generalised_epilepsy                                                                 0 non-null      float64
 6   Jones_et_al_2016_Chronotype                                                                       8807 non-null   float64
 7   Jones_et_al_2016_SleepDuration                                                                    8807 non-null   float64
 8   MDD_MHQ_BIP_METACARPA_INFO6_A5_NTOT_no23andMe_noUKBB.txt.gz                                       7801 non-null   float64
 9   MDD_MHQ_METACARPA_INFO6_A5_NTOT_no23andMe_noUKBB.txt.gz                                           7826 non-null   float64
 10  MHQ_Depression_WG_MAF1_INFO4_HRC_Only_Filtered_Dups_FOR_METACARPA_INFO6_A5_NTOT.txt.gz            7977 non-null   float64
 11  MHQ_Recurrent_Depression_WG_MAF1_INFO4_HRC_Only_Filtered_Dups_FOR_METACARPA_INFO6_A5_NTOT.txt.gz  7971 non-null   float64
 12  MHQ_Single_Depression_WG_MAF1_INFO4_HRC_Only_Filtered_Dups_FOR_METACARPA_INFO6_A5_NTOT.txt.gz     7972 non-null   float64
 13  MHQ_Subthreshold_WG_MAF1_INFO4_HRC_Only_Filtered_Dups_FOR_METACARPA_INFO6_A5_NTOT.txt.gz          7973 non-null   float64
 14  PGC3_SCZ_wave3_public.v2.txt.gz                                                                   9707 non-null   float64
 15  PGC_ADHD_EUR_2017                                                                                 7880 non-null   float64
 16  PGC_ASD_2017_CEU                                                                                  7287 non-null   float64
 17  SSGAC_Depressive_Symptoms                                                                         7888 non-null   float64
 18  SSGAC_Education_Years_Pooled                                                                      8824 non-null   float64
 19  UKB_1160_Sleep_duration                                                                           8829 non-null   float64
 20  UKB_1180_Morning_or_evening_person_chronotype                                                     8829 non-null   float64
 21  UKB_1200_Sleeplessness_or_insomnia                                                                8829 non-null   float64
 22  UKB_20002_1243_self_reported_psychological_or_psychiatric_problem                                 8829 non-null   float64
 23  UKB_20002_1262_self_reported_parkinsons_disease                                                   8829 non-null   float64
 24  UKB_20002_1265_self_reported_migraine                                                             8829 non-null   float64
 25  UKB_20002_1289_self_reported_schizophrenia                                                        8829 non-null   float64
 26  UKB_20002_1616_self_reported_insomnia                                                             8829 non-null   float64
 27  UKB_20016_Fluid_intelligence_score                                                                8829 non-null   float64
 28  UKB_20127_Neuroticism_score                                                                       8829 non-null   float64
 29  UKB_G40_Diagnoses_main_ICD10_G40_Epilepsy                                                         8829 non-null   float64
 30  UKB_G43_Diagnoses_main_ICD10_G43_Migraine                                                         8829 non-null   float64
 31  anxiety.meta.full.cc.txt.gz                                                                       8449 non-null   float64
 32  anxiety.meta.full.fs.txt.gz                                                                       8418 non-null   float64
 33  daner_PGC_BIP32b_mds7a_0416a.txt.gz                                                               9988 non-null   float64
 34  daner_PGC_BIP32b_mds7a_mds7a_BD1.0416a_INFO6_A5_NTOT.txt.gz                                       8108 non-null   float64
 35  daner_PGC_BIP32b_mds7a_mds7a_BD2.0416a_INFO6_A5_NTOT.txt.gz                                       8096 non-null   float64
 36  daner_adhd_meta_filtered_NA_iPSYCH23_PGC11_sigPCs_woSEX_2ell6sd_EUR_Neff_70.txt.gz                8795 non-null   float64
 37  iPSYCH-PGC_ASD_Nov2017.txt.gz                                                                     9654 non-null   float64
 38  ieu-a-1000                                                                                        8792 non-null   float64
 39  ieu-a-1029                                                                                        4180 non-null   float64
 40  ieu-a-1041                                                                                        8494 non-null   float64
 41  ieu-a-1042                                                                                        8492 non-null   float64
 42  ieu-a-1043                                                                                        8496 non-null   float64
 43  ieu-a-1044                                                                                        8495 non-null   float64
 44  ieu-a-1045                                                                                        8496 non-null   float64
 45  ieu-a-1046                                                                                        8491 non-null   float64
 46  ieu-a-1047                                                                                        8496 non-null   float64
 47  ieu-a-1048                                                                                        8498 non-null   float64
 48  ieu-a-1061                                                                                        4117 non-null   float64
 49  ieu-a-1063                                                                                        4059 non-null   float64
 50  ieu-a-1064                                                                                        4120 non-null   float64
 51  ieu-a-1065                                                                                        4117 non-null   float64
 52  ieu-a-1066                                                                                        4120 non-null   float64
 53  ieu-a-1067                                                                                        4116 non-null   float64
 54  ieu-a-1068                                                                                        4117 non-null   float64
 55  ieu-a-1085                                                                                        9043 non-null   float64
 56  ieu-a-118                                                                                         8939 non-null   float64
 57  ieu-a-297                                                                                         8870 non-null   float64
 58  ieu-a-990                                                                                         7676 non-null   float64
 59  ieu-b-10                                                                                          5586 non-null   float64
 60  ieu-b-11                                                                                          5817 non-null   float64
 61  ieu-b-12                                                                                          5828 non-null   float64
 62  ieu-b-13                                                                                          5820 non-null   float64
 63  ieu-b-14                                                                                          5780 non-null   float64
 64  ieu-b-15                                                                                          5821 non-null   float64
 65  ieu-b-16                                                                                          5829 non-null   float64
 66  ieu-b-17                                                                                          5814 non-null   float64
 67  ieu-b-18                                                                                          7781 non-null   float64
 68  ieu-b-2                                                                                           9360 non-null   float64
 69  ieu-b-5070                                                                                        7585 non-null   float64
 70  ieu-b-7                                                                                           9227 non-null   float64
 71  ieu-b-8                                                                                           5508 non-null   float64
 72  ieu-b-9                                                                                           5631 non-null   float64
 73  ocd_aug2017.txt.gz                                                                                9774 non-null   float64
 74  pgc-bip2021-BDI.vcf.txt.gz                                                                        9551 non-null   float64
 75  pgc-bip2021-BDII.vcf.txt.gz                                                                       9418 non-null   float64
 76  pgc-bip2021-all.vcf.txt.gz                                                                        9718 non-null   float64
 77  pgc.scz2                                                                                          8760 non-null   float64
 78  pgcAN2.2019-07.vcf.txt.gz                                                                         9057 non-null   float64
 79  pts_all_freeze2_overall.txt.gz                                                                    9988 non-null   float64
dtypes: float64(80)
memory usage: 6.2 MB</code></pre>
</div>
</div>
</section>
</section>
<section id="remove-studies-with-unreliable-standard-errors" class="level1">
<h1>Remove studies with unreliable standard errors</h1>
<p>To eliminate unreliable estimates of genetic associations, we remove samples (phenotypes) whose median standard error (SE) of beta value or log odds ratio is more than 0.2. The median SE for all the phenotypes are shown in <a href="#fig-se-vs-beta-std">Figure&nbsp;3</a></p>
<p>Note: For SE, the mean is not optimal because, in some studies, there are few SNPs with very high (~10000) SE, which inflates the mean.</p>
<div class="cell" data-execution_count="32">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>mean_se  <span class="op">=</span> se_df_sub.median(axis <span class="op">=</span> <span class="dv">0</span>, skipna <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>mean_se  <span class="op">=</span> pd.DataFrame(mean_se).set_axis([<span class="st">"mean_se"</span>], axis <span class="op">=</span> <span class="dv">1</span>)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>beta_std <span class="op">=</span> beta_df_sub.std(axis <span class="op">=</span> <span class="dv">0</span>, skipna <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>beta_std <span class="op">=</span> pd.DataFrame(beta_std).set_axis([<span class="st">"beta_std"</span>], axis <span class="op">=</span> <span class="dv">1</span>)</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>error_df <span class="op">=</span> pd.concat([mean_se, beta_std], axis <span class="op">=</span> <span class="dv">1</span>)</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure()</span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>ax1 <span class="op">=</span> fig.add_subplot(<span class="dv">111</span>)</span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a>ax1.scatter(np.log10(error_df[<span class="st">'beta_std'</span>]), np.log10(error_df[<span class="st">'mean_se'</span>]), alpha <span class="op">=</span> <span class="fl">0.5</span>, s <span class="op">=</span> <span class="dv">100</span>)</span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a>mpl_utils.set_xticks(ax1, scale <span class="op">=</span> <span class="st">'log10'</span>, spacing <span class="op">=</span> <span class="st">'log10'</span>)</span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a>mpl_utils.set_yticks(ax1, scale <span class="op">=</span> <span class="st">'log10'</span>, spacing <span class="op">=</span> <span class="st">'log10'</span>)</span>
<span id="cb46-12"><a href="#cb46-12" aria-hidden="true" tabindex="-1"></a>mpl_utils.plot_diag(ax1)</span>
<span id="cb46-13"><a href="#cb46-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-14"><a href="#cb46-14" aria-hidden="true" tabindex="-1"></a>keep_columns <span class="op">=</span> error_df.query(<span class="st">"mean_se &lt;= 0.2"</span>).index</span>
<span id="cb46-15"><a href="#cb46-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> pid <span class="kw">in</span> error_df.index.to_list():</span>
<span id="cb46-16"><a href="#cb46-16" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> pid <span class="kw">not</span> <span class="kw">in</span> keep_columns:</span>
<span id="cb46-17"><a href="#cb46-17" aria-hidden="true" tabindex="-1"></a>        pid_text <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>pid<span class="sc">}</span><span class="ss"> / </span><span class="sc">{</span>phenotype_dict[pid]<span class="sc">}</span><span class="ss">"</span></span>
<span id="cb46-18"><a href="#cb46-18" aria-hidden="true" tabindex="-1"></a>        xval <span class="op">=</span> np.log10(error_df.loc[pid][<span class="st">'beta_std'</span>])</span>
<span id="cb46-19"><a href="#cb46-19" aria-hidden="true" tabindex="-1"></a>        yval <span class="op">=</span> np.log10(error_df.loc[pid][<span class="st">'mean_se'</span>])</span>
<span id="cb46-20"><a href="#cb46-20" aria-hidden="true" tabindex="-1"></a>        ax1.annotate(pid_text, (xval, yval))</span>
<span id="cb46-21"><a href="#cb46-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-22"><a href="#cb46-22" aria-hidden="true" tabindex="-1"></a>ax1.set_xlabel(<span class="vs">r"Standard Deviation of mean $\beta$"</span>)</span>
<span id="cb46-23"><a href="#cb46-23" aria-hidden="true" tabindex="-1"></a>ax1.set_ylabel(<span class="vs">r"Median of SE"</span>)</span>
<span id="cb46-24"><a href="#cb46-24" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div id="fig-se-vs-beta-std" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="2023-05-12-preprocess_files/figure-html/fig-se-vs-beta-std-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;3: Calibration of SE against std of beta</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell" data-execution_count="33">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Keeping </span><span class="sc">{</span><span class="bu">len</span>(keep_columns)<span class="sc">}</span><span class="ss"> phenotypes with low standard error"</span>)</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>beta_df_sub_lowse   <span class="op">=</span> beta_df_sub[keep_columns]</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>prec_df_sub_lowse   <span class="op">=</span> prec_df_sub[keep_columns]</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>zscore_df_sub_lowse <span class="op">=</span> zscore_df_sub[keep_columns]</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>se_df_sub_lowse     <span class="op">=</span> se_df_sub[keep_columns]</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>remove_columns <span class="op">=</span> [x <span class="cf">for</span> x <span class="kw">in</span> error_df.index.to_list() <span class="cf">if</span> x <span class="kw">not</span> <span class="kw">in</span> keep_columns]</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Phenotypes removed:"</span>)</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> x <span class="kw">in</span> remove_columns:</span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>    se <span class="op">=</span> error_df.loc[x][<span class="st">'mean_se'</span>]</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>    bstd <span class="op">=</span> error_df.loc[x][<span class="st">'beta_std'</span>]</span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="sc">{</span>se<span class="sc">:.3f}</span><span class="ch">\t</span><span class="sc">{</span>bstd<span class="sc">:.3f}</span><span class="ch">\t</span><span class="sc">{</span>x<span class="sc">}</span><span class="ch">\t</span><span class="sc">{</span>phenotype_dict[x]<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Keeping 70 phenotypes with low standard error
Phenotypes removed:
    2322.222    4272.766    ENIGMA_Intracraneal_Volume  Volume
    nan nan ILAE_Genetic_generalised_epilepsy   Epilepsy
    2330.240    4287.398    ieu-a-1041  Volume
    1.326   2.479   ieu-a-1042  Volume
    2.855   5.450   ieu-a-1043  Volume
    5.930   11.025  ieu-a-1044  Volume
    5.691   10.837  ieu-a-1045  Volume
    2.319   4.306   ieu-a-1046  Volume
    7.255   13.720  ieu-a-1047  Volume
    7.590   14.253  ieu-a-1048  Volume</code></pre>
</div>
</div>
</section>
<section id="remove-studies-with-unreliable-z-scores" class="level1">
<h1>Remove studies with unreliable z-scores</h1>
<p>To eliminate unreliable z-scores, we make sure that the median of z-scores is less than 0.2</p>
<div class="cell" data-execution_count="34">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>median_zscore <span class="op">=</span> zscore_df_sub_lowse.median(axis <span class="op">=</span> <span class="dv">0</span>, skipna <span class="op">=</span> <span class="va">True</span>).<span class="bu">abs</span>()</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>median_zscore  <span class="op">=</span> pd.DataFrame(median_zscore).set_axis([<span class="st">"median_zscore"</span>], axis <span class="op">=</span> <span class="dv">1</span>)</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>keep_columns <span class="op">=</span> median_zscore.query(<span class="st">"median_zscore &lt;= 0.2"</span>).index</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>remove_columns <span class="op">=</span> median_zscore.query(<span class="st">"median_zscore &gt; 0.2"</span>).index</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Keeping </span><span class="sc">{</span><span class="bu">len</span>(keep_columns)<span class="sc">}</span><span class="ss"> phenotypes with reasonable z-scores."</span>)</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>beta_df_sub_lowse_zchi2   <span class="op">=</span> beta_df_sub_lowse[keep_columns]</span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>prec_df_sub_lowse_zchi2   <span class="op">=</span> prec_df_sub_lowse[keep_columns]</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>zscore_df_sub_lowse_zchi2 <span class="op">=</span> zscore_df_sub_lowse[keep_columns]</span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a>se_df_sub_lowse_zchi2     <span class="op">=</span> se_df_sub_lowse[keep_columns]</span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Phenotypes removed:"</span>)</span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> x <span class="kw">in</span> remove_columns:</span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\t</span><span class="sc">{</span>x<span class="sc">}</span><span class="ch">\t</span><span class="sc">{</span>phenotype_dict[x]<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Keeping 69 phenotypes with reasonable z-scores.
Phenotypes removed:
    ieu-b-5070  SZ</code></pre>
</div>
</div>
<div class="cell" data-execution_count="35">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> q1(x, axis <span class="op">=</span> <span class="va">None</span>):</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> np.percentile(x, <span class="dv">25</span>, axis <span class="op">=</span> axis)</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> q3(x, axis <span class="op">=</span> <span class="va">None</span>):</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> np.percentile(x, <span class="dv">75</span>, axis <span class="op">=</span> axis)</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> iqr_outlier(x, axis <span class="op">=</span> <span class="va">None</span>, bar <span class="op">=</span> <span class="fl">1.5</span>, side <span class="op">=</span> <span class="st">'both'</span>):</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> side <span class="kw">in</span> [<span class="st">'gt'</span>, <span class="st">'lt'</span>, <span class="st">'both'</span>], <span class="st">'Side should be `gt`, `lt` or `both`.'</span></span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a>    d_iqr <span class="op">=</span> sc_stats.iqr(x, axis <span class="op">=</span> axis)</span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a>    d_q1 <span class="op">=</span> q1(x, axis <span class="op">=</span> axis)</span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a>    d_q3 <span class="op">=</span> q3(x, axis <span class="op">=</span> axis)</span>
<span id="cb51-13"><a href="#cb51-13" aria-hidden="true" tabindex="-1"></a>    iqr_distance <span class="op">=</span> np.multiply(d_iqr, bar)</span>
<span id="cb51-14"><a href="#cb51-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-15"><a href="#cb51-15" aria-hidden="true" tabindex="-1"></a>    stat_shape <span class="op">=</span> <span class="bu">list</span>(x.shape)</span>
<span id="cb51-16"><a href="#cb51-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-17"><a href="#cb51-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">isinstance</span>(axis, collections.abc.Iterable):</span>
<span id="cb51-18"><a href="#cb51-18" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> single_axis <span class="kw">in</span> axis:</span>
<span id="cb51-19"><a href="#cb51-19" aria-hidden="true" tabindex="-1"></a>            stat_shape[single_axis] <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb51-20"><a href="#cb51-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb51-21"><a href="#cb51-21" aria-hidden="true" tabindex="-1"></a>        stat_shape[axis] <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb51-22"><a href="#cb51-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-23"><a href="#cb51-23" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> side <span class="kw">in</span> [<span class="st">'gt'</span>, <span class="st">'both'</span>]:</span>
<span id="cb51-24"><a href="#cb51-24" aria-hidden="true" tabindex="-1"></a>        upper_range <span class="op">=</span> d_q3 <span class="op">+</span> iqr_distance</span>
<span id="cb51-25"><a href="#cb51-25" aria-hidden="true" tabindex="-1"></a>        upper_outlier <span class="op">=</span> np.greater(x <span class="op">-</span> upper_range.reshape(stat_shape), <span class="dv">0</span>)</span>
<span id="cb51-26"><a href="#cb51-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> side <span class="kw">in</span> [<span class="st">'lt'</span>, <span class="st">'both'</span>]:</span>
<span id="cb51-27"><a href="#cb51-27" aria-hidden="true" tabindex="-1"></a>        lower_range <span class="op">=</span> d_q1 <span class="op">-</span> iqr_distance</span>
<span id="cb51-28"><a href="#cb51-28" aria-hidden="true" tabindex="-1"></a>        lower_outlier <span class="op">=</span> np.less(x <span class="op">-</span> lower_range.reshape(stat_shape), <span class="dv">0</span>)</span>
<span id="cb51-29"><a href="#cb51-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-30"><a href="#cb51-30" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> side <span class="op">==</span> <span class="st">'gt'</span>:</span>
<span id="cb51-31"><a href="#cb51-31" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> upper_outlier</span>
<span id="cb51-32"><a href="#cb51-32" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> side <span class="op">==</span> <span class="st">'lt'</span>:</span>
<span id="cb51-33"><a href="#cb51-33" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> lower_outlier</span>
<span id="cb51-34"><a href="#cb51-34" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> side <span class="op">==</span> <span class="st">'both'</span>:</span>
<span id="cb51-35"><a href="#cb51-35" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> np.logical_or(upper_outlier, lower_outlier)</span>
<span id="cb51-36"><a href="#cb51-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-37"><a href="#cb51-37" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_density(x, data):</span>
<span id="cb51-38"><a href="#cb51-38" aria-hidden="true" tabindex="-1"></a>    density <span class="op">=</span> sc_stats.gaussian_kde(data)</span>
<span id="cb51-39"><a href="#cb51-39" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> density.pdf(x)</span>
<span id="cb51-40"><a href="#cb51-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-41"><a href="#cb51-41" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_bins(data, nbin, xmin, xmax):</span>
<span id="cb51-42"><a href="#cb51-42" aria-hidden="true" tabindex="-1"></a>    xdelta <span class="op">=</span> (np.<span class="bu">max</span>(data) <span class="op">-</span> np.<span class="bu">min</span>(data)) <span class="op">/</span> <span class="dv">10</span></span>
<span id="cb51-43"><a href="#cb51-43" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> xmin: xmin <span class="op">=</span> np.<span class="bu">min</span>(data) <span class="op">-</span> xdelta</span>
<span id="cb51-44"><a href="#cb51-44" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> xmax: xmax <span class="op">=</span> np.<span class="bu">max</span>(data) <span class="op">+</span> xdelta</span>
<span id="cb51-45"><a href="#cb51-45" aria-hidden="true" tabindex="-1"></a>    bins <span class="op">=</span> np.linspace(xmin, xmax, nbin)</span>
<span id="cb51-46"><a href="#cb51-46" aria-hidden="true" tabindex="-1"></a>    xbin <span class="op">=</span> [(bins[i] <span class="op">+</span> bins[i<span class="op">+</span><span class="dv">1</span>]) <span class="op">/</span> <span class="dv">2</span> <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(bins.shape[<span class="dv">0</span>] <span class="op">-</span> <span class="dv">1</span>)] <span class="co"># centers of the bins</span></span>
<span id="cb51-47"><a href="#cb51-47" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> xmin, xmax, bins, xbin</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-execution_count="36">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure()</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>ax1 <span class="op">=</span> fig.add_subplot(<span class="dv">111</span>)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>remove_columns <span class="op">=</span> []</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> pid <span class="kw">in</span> zscore_df_sub_lowse_zchi2.columns.to_list():</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>    x <span class="op">=</span> zscore_df_sub_lowse_zchi2[pid].values</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>    x_dropna <span class="op">=</span> x[<span class="op">~</span>np.isnan(x)]</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>    outlier_mask <span class="op">=</span> iqr_outlier(x_dropna, axis <span class="op">=</span> <span class="dv">0</span>, bar <span class="op">=</span> <span class="dv">5</span>)</span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>    <span class="co">#data = np.square(x_dropna[~outlier_mask])</span></span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>    data <span class="op">=</span> x_dropna[<span class="op">~</span>outlier_mask]</span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a>    xmin, xmax, bins, xbin <span class="op">=</span> get_bins(data, <span class="dv">100</span>, <span class="va">None</span>, <span class="va">None</span>)</span>
<span id="cb52-12"><a href="#cb52-12" aria-hidden="true" tabindex="-1"></a>    curve <span class="op">=</span> get_density(xbin, data)</span>
<span id="cb52-13"><a href="#cb52-13" aria-hidden="true" tabindex="-1"></a>    <span class="co">#print (pid, xmax)</span></span>
<span id="cb52-14"><a href="#cb52-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> xmax <span class="op">&gt;=</span> <span class="dv">1000</span>:</span>
<span id="cb52-15"><a href="#cb52-15" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span> (pid, xmax)</span>
<span id="cb52-16"><a href="#cb52-16" aria-hidden="true" tabindex="-1"></a>        remove_columns.append(pid)</span>
<span id="cb52-17"><a href="#cb52-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb52-18"><a href="#cb52-18" aria-hidden="true" tabindex="-1"></a>        ax1.plot(xbin, curve, label <span class="op">=</span> pid)</span>
<span id="cb52-19"><a href="#cb52-19" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb52-20"><a href="#cb52-20" aria-hidden="true" tabindex="-1"></a><span class="co">#ax1.legend()</span></span>
<span id="cb52-21"><a href="#cb52-21" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div id="fig-zscore-distribution" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="2023-05-12-preprocess_files/figure-html/fig-zscore-distribution-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;4: Distribution of z-scores for all phenotypes</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="flip-signs" class="level1">
<h1>Flip Signs</h1>
<div class="cell" data-execution_count="37">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>to_flip <span class="op">=</span> [</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"AD_sumstats_Jansenetal_2019sept.txt.gz"</span>,</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"daner_adhd_meta_filtered_NA_iPSYCH23_PGC11_sigPCs_woSEX_2ell6sd_EUR_Neff_70.txt.gz"</span>,</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Jones_et_al_2016_Chronotype"</span>,</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Jones_et_al_2016_SleepDuration"</span>,</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"UKB_1160_Sleep_duration"</span>,</span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>    <span class="co">#"UKB_1180_Morning_or_evening_person_chronotype",</span></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"MDD_MHQ_BIP_METACARPA_INFO6_A5_NTOT_no23andMe_noUKBB.txt.gz"</span>,</span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"MDD_MHQ_METACARPA_INFO6_A5_NTOT_no23andMe_noUKBB.txt.gz"</span>,</span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"PGC3_SCZ_wave3_public.v2.txt.gz"</span>,</span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-13"><a href="#cb53-13" aria-hidden="true" tabindex="-1"></a>remaining_columns <span class="op">=</span> beta_df_sub_lowse_zchi2.columns.to_list()</span>
<span id="cb53-14"><a href="#cb53-14" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> flip_id <span class="kw">in</span> to_flip:</span>
<span id="cb53-15"><a href="#cb53-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> flip_id <span class="kw">in</span> remaining_columns:</span>
<span id="cb53-16"><a href="#cb53-16" aria-hidden="true" tabindex="-1"></a>        beta_df_sub_lowse_zchi2.loc[:, (flip_id)] <span class="op">=</span> <span class="op">-</span> beta_df_sub_lowse_zchi2[flip_id]</span>
<span id="cb53-17"><a href="#cb53-17" aria-hidden="true" tabindex="-1"></a>        zscore_df_sub_lowse_zchi2.loc[:, (flip_id)] <span class="op">=</span> <span class="op">-</span> zscore_df_sub_lowse_zchi2[flip_id]</span>
<span id="cb53-18"><a href="#cb53-18" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(flip_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>AD_sumstats_Jansenetal_2019sept.txt.gz
daner_adhd_meta_filtered_NA_iPSYCH23_PGC11_sigPCs_woSEX_2ell6sd_EUR_Neff_70.txt.gz
Jones_et_al_2016_Chronotype
Jones_et_al_2016_SleepDuration
UKB_1160_Sleep_duration
MDD_MHQ_BIP_METACARPA_INFO6_A5_NTOT_no23andMe_noUKBB.txt.gz
MDD_MHQ_METACARPA_INFO6_A5_NTOT_no23andMe_noUKBB.txt.gz
PGC3_SCZ_wave3_public.v2.txt.gz</code></pre>
</div>
</div>
</section>
<section id="save-pickle" class="level1">
<h1>Save Pickle</h1>
<div class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>beta_df_filename <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>data_dir<span class="sc">}</span><span class="ss">/beta_df.pkl"</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>prec_df_filename <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>data_dir<span class="sc">}</span><span class="ss">/prec_df.pkl"</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>se_df_filename   <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>data_dir<span class="sc">}</span><span class="ss">/se_df.pkl"</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>zscore_df_filename <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>data_dir<span class="sc">}</span><span class="ss">/zscore_df.pkl"</span></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>beta_df_sub_lowse_zchi2.to_pickle(beta_df_filename)</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>prec_df_sub_lowse_zchi2.to_pickle(prec_df_filename)</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>se_df_sub_lowse_zchi2.to_pickle(se_df_filename)</span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a>zscore_df_sub_lowse_zchi2.to_pickle(zscore_df_filename)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="39">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>zscore_df_sub_lowse_zchi2.columns.to_list()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="39">
<pre><code>['AD_sumstats_Jansenetal_2019sept.txt.gz',
 'CNCR_Insomnia_all',
 'GPC-NEO-NEUROTICISM',
 'IGAP_Alzheimer',
 'Jones_et_al_2016_Chronotype',
 'Jones_et_al_2016_SleepDuration',
 'MDD_MHQ_BIP_METACARPA_INFO6_A5_NTOT_no23andMe_noUKBB.txt.gz',
 'MDD_MHQ_METACARPA_INFO6_A5_NTOT_no23andMe_noUKBB.txt.gz',
 'MHQ_Depression_WG_MAF1_INFO4_HRC_Only_Filtered_Dups_FOR_METACARPA_INFO6_A5_NTOT.txt.gz',
 'MHQ_Recurrent_Depression_WG_MAF1_INFO4_HRC_Only_Filtered_Dups_FOR_METACARPA_INFO6_A5_NTOT.txt.gz',
 'MHQ_Single_Depression_WG_MAF1_INFO4_HRC_Only_Filtered_Dups_FOR_METACARPA_INFO6_A5_NTOT.txt.gz',
 'MHQ_Subthreshold_WG_MAF1_INFO4_HRC_Only_Filtered_Dups_FOR_METACARPA_INFO6_A5_NTOT.txt.gz',
 'PGC3_SCZ_wave3_public.v2.txt.gz',
 'PGC_ADHD_EUR_2017',
 'PGC_ASD_2017_CEU',
 'SSGAC_Depressive_Symptoms',
 'SSGAC_Education_Years_Pooled',
 'UKB_1160_Sleep_duration',
 'UKB_1180_Morning_or_evening_person_chronotype',
 'UKB_1200_Sleeplessness_or_insomnia',
 'UKB_20002_1243_self_reported_psychological_or_psychiatric_problem',
 'UKB_20002_1262_self_reported_parkinsons_disease',
 'UKB_20002_1265_self_reported_migraine',
 'UKB_20002_1289_self_reported_schizophrenia',
 'UKB_20002_1616_self_reported_insomnia',
 'UKB_20016_Fluid_intelligence_score',
 'UKB_20127_Neuroticism_score',
 'UKB_G40_Diagnoses_main_ICD10_G40_Epilepsy',
 'UKB_G43_Diagnoses_main_ICD10_G43_Migraine',
 'anxiety.meta.full.cc.txt.gz',
 'anxiety.meta.full.fs.txt.gz',
 'daner_PGC_BIP32b_mds7a_0416a.txt.gz',
 'daner_PGC_BIP32b_mds7a_mds7a_BD1.0416a_INFO6_A5_NTOT.txt.gz',
 'daner_PGC_BIP32b_mds7a_mds7a_BD2.0416a_INFO6_A5_NTOT.txt.gz',
 'daner_adhd_meta_filtered_NA_iPSYCH23_PGC11_sigPCs_woSEX_2ell6sd_EUR_Neff_70.txt.gz',
 'iPSYCH-PGC_ASD_Nov2017.txt.gz',
 'ieu-a-1000',
 'ieu-a-1029',
 'ieu-a-1061',
 'ieu-a-1063',
 'ieu-a-1064',
 'ieu-a-1065',
 'ieu-a-1066',
 'ieu-a-1067',
 'ieu-a-1068',
 'ieu-a-1085',
 'ieu-a-118',
 'ieu-a-297',
 'ieu-a-990',
 'ieu-b-10',
 'ieu-b-11',
 'ieu-b-12',
 'ieu-b-13',
 'ieu-b-14',
 'ieu-b-15',
 'ieu-b-16',
 'ieu-b-17',
 'ieu-b-18',
 'ieu-b-2',
 'ieu-b-7',
 'ieu-b-8',
 'ieu-b-9',
 'ocd_aug2017.txt.gz',
 'pgc-bip2021-BDI.vcf.txt.gz',
 'pgc-bip2021-BDII.vcf.txt.gz',
 'pgc-bip2021-all.vcf.txt.gz',
 'pgc.scz2',
 'pgcAN2.2019-07.vcf.txt.gz',
 'pts_all_freeze2_overall.txt.gz']</code></pre>
</div>
</div>
</section>
<section id="save-snp-information" class="level1">
<h1>Save SNP information</h1>
<div class="cell" data-execution_count="40">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>assoc_df_fa_nsnp_nodup</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="40">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">SNP</th>
<th data-quarto-table-cell-role="th">A1</th>
<th data-quarto-table-cell-role="th">A2</th>
<th data-quarto-table-cell-role="th">CHR</th>
<th data-quarto-table-cell-role="th">BP</th>
<th data-quarto-table-cell-role="th">Z</th>
<th data-quarto-table-cell-role="th">P</th>
<th data-quarto-table-cell-role="th">BETA</th>
<th data-quarto-table-cell-role="th">SE</th>
<th data-quarto-table-cell-role="th">ID</th>
<th data-quarto-table-cell-role="th">TRAIT</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>rs10486722</td>
<td>C</td>
<td>T</td>
<td>chr7</td>
<td>41772310</td>
<td>0.881222</td>
<td>0.378198</td>
<td>0.001757</td>
<td>0.001994</td>
<td>UKB_1160_Sleep_duration</td>
<td>UKB_1160_Sleep_duration</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>rs10486722</td>
<td>C</td>
<td>T</td>
<td>chr7</td>
<td>41772310</td>
<td>1.369354</td>
<td>0.170889</td>
<td>0.003476</td>
<td>0.002538</td>
<td>UKB_1180_Morning_or_evening_person_chronotype</td>
<td>UKB_1180_Morning_or_evening_person_chronotype</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>rs10486722</td>
<td>C</td>
<td>T</td>
<td>chr7</td>
<td>41772310</td>
<td>-0.138782</td>
<td>0.889622</td>
<td>-0.000257</td>
<td>0.001850</td>
<td>UKB_1200_Sleeplessness_or_insomnia</td>
<td>UKB_1200_Sleeplessness_or_insomnia</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>rs10486722</td>
<td>C</td>
<td>T</td>
<td>chr7</td>
<td>41772310</td>
<td>0.442336</td>
<td>0.658246</td>
<td>0.000035</td>
<td>0.000080</td>
<td>UKB_20002_1243_self_reported_psychological_or_...</td>
<td>UKB_20002_1243_self_reported_psychological_or_...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>rs10486722</td>
<td>C</td>
<td>T</td>
<td>chr7</td>
<td>41772310</td>
<td>0.275442</td>
<td>0.782977</td>
<td>0.000030</td>
<td>0.000110</td>
<td>UKB_20002_1262_self_reported_parkinsons_disease</td>
<td>UKB_20002_1262_self_reported_parkinsons_disease</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">676256</td>
<td>rs2322</td>
<td>C</td>
<td>G</td>
<td>4</td>
<td>9701603</td>
<td>NaN</td>
<td>0.396100</td>
<td>-0.077600</td>
<td>0.091500</td>
<td>ieu-b-7</td>
<td>Parkinson's disease</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">676257</td>
<td>rs2322</td>
<td>C</td>
<td>G</td>
<td>4</td>
<td>9701603</td>
<td>NaN</td>
<td>0.107599</td>
<td>0.024041</td>
<td>0.014941</td>
<td>MHQ_Depression_WG_MAF1_INFO4_HRC_Only_Filtered...</td>
<td>MHQ_Depression_WG_MAF1_INFO4_HRC_Only_Filtered...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">676258</td>
<td>rs2322</td>
<td>C</td>
<td>G</td>
<td>4</td>
<td>9701603</td>
<td>NaN</td>
<td>0.430338</td>
<td>0.016141</td>
<td>0.020468</td>
<td>MHQ_Recurrent_Depression_WG_MAF1_INFO4_HRC_Onl...</td>
<td>MHQ_Recurrent_Depression_WG_MAF1_INFO4_HRC_Onl...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">676259</td>
<td>rs2322</td>
<td>C</td>
<td>G</td>
<td>4</td>
<td>9701603</td>
<td>NaN</td>
<td>0.068770</td>
<td>0.049301</td>
<td>0.027090</td>
<td>MHQ_Single_Depression_WG_MAF1_INFO4_HRC_Only_F...</td>
<td>MHQ_Single_Depression_WG_MAF1_INFO4_HRC_Only_F...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">676260</td>
<td>rs2322</td>
<td>C</td>
<td>G</td>
<td>4</td>
<td>9701603</td>
<td>NaN</td>
<td>0.721340</td>
<td>0.006329</td>
<td>0.017746</td>
<td>MHQ_Subthreshold_WG_MAF1_INFO4_HRC_Only_Filter...</td>
<td>MHQ_Subthreshold_WG_MAF1_INFO4_HRC_Only_Filter...</td>
</tr>
</tbody>
</table>

<p>676229 rows × 11 columns</p>
</div>
</div>
</div>
<div class="cell" data-execution_count="72">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>all_snp_info <span class="op">=</span> assoc_df_fa_nsnp_nodup[[<span class="st">'SNP'</span>, <span class="st">'CHR'</span>, <span class="st">'BP'</span>]].copy()</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>all_snp_info_nona <span class="op">=</span> all_snp_info.dropna(axis <span class="op">=</span> <span class="dv">0</span>)</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>all_snp_info_nona.loc[:, <span class="st">'CHR'</span>] <span class="op">=</span> all_snp_info_nona.CHR.<span class="bu">str</span>.replace(<span class="st">'chr'</span>, <span class="st">''</span>)</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>all_snp_info_nodup <span class="op">=</span> all_snp_info_nona.drop_duplicates(subset<span class="op">=</span>[<span class="st">'SNP'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-execution_count="73">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>all_snp_info_nodup</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="73">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">SNP</th>
<th data-quarto-table-cell-role="th">CHR</th>
<th data-quarto-table-cell-role="th">BP</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>rs10486722</td>
<td>7</td>
<td>41772310</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">91</td>
<td>rs2416745</td>
<td>9</td>
<td>120018976</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">182</td>
<td>rs9287971</td>
<td>2</td>
<td>174067024</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">273</td>
<td>rs12660608</td>
<td>6</td>
<td>123017369</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">364</td>
<td>rs1693523</td>
<td>15</td>
<td>52535142</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">676161</td>
<td>rs7199470</td>
<td>16</td>
<td>10137879</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">676181</td>
<td>rs77337722</td>
<td>18</td>
<td>74935886</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">676201</td>
<td>rs183211079</td>
<td>2</td>
<td>151553344</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">676221</td>
<td>rs78346960</td>
<td>17</td>
<td>9121155</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">676241</td>
<td>rs2322</td>
<td>4</td>
<td>9701603</td>
</tr>
</tbody>
</table>

<p>10068 rows × 3 columns</p>
</div>
</div>
</div>
<div class="cell" data-execution_count="74">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>snp_info_filename <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>data_dir<span class="sc">}</span><span class="ss">/snp_info.pkl"</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>all_snp_info_nodup.to_pickle(snp_info_filename)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("^(?:http:|https:)\/\/banskt\.github\.io\/npd-notes\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
          // default icon
          link.classList.add("external");
      }
    }
});
</script>
</div> <!-- /content -->



</body></html>