{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "da6c7e3d-06d3-4be8-97f2-31b9365baffe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "263549c1-830e-4056-b7cf-4d106cb9a3cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = \"/gpfs/commons/home/sbanerjee/work/npd/PanUKB/data\"\n",
    "result_dir = \"/gpfs/commons/home/sbanerjee/work/npd/PanUKB/results/nnsparsh/noRx\"\n",
    "variant_dir = \"/gpfs/commons/home/sbanerjee/npddata/panukb/ldsc/varinfo\"\n",
    "\n",
    "zscore_df = pd.read_pickle(os.path.join(data_dir, f\"modselect/zscore_noRx.pkl\"))\n",
    "# trait_df  = pd.read_pickle(os.path.join(data_dir, f\"modselect/traits_noRx.pkl\"))\n",
    "trait_df  = pd.read_pickle(os.path.join(data_dir, f\"modselect/traits_all_with_desc.pkl\"))\n",
    "\n",
    "variant_filename = f\"{data_dir}/allvar.pruned.closesttss.hugo\"\n",
    "variant_df = pd.read_csv(variant_filename, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ee84bda0-5d5b-403d-9999-8b87466f5e04",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df_selected = variant_df.loc[zscore_df.index.to_numpy()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d26c17c1-59e8-498a-af79-37354e7888de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>rsid</th>\n",
       "      <th>varid</th>\n",
       "      <th>maf_EUR</th>\n",
       "      <th>name_1kg</th>\n",
       "      <th>tie</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>name</th>\n",
       "      <th>omim_id</th>\n",
       "      <th>notfound</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs6657440</td>\n",
       "      <td>1</td>\n",
       "      <td>850780</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs6657440</td>\n",
       "      <td>rs6657440</td>\n",
       "      <td>0.395650</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>1</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>HGNC:28706</td>\n",
       "      <td>sterile alpha motif domain containing 11</td>\n",
       "      <td>616765.0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs7418179</td>\n",
       "      <td>1</td>\n",
       "      <td>858801</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs7418179</td>\n",
       "      <td>1:858801_A_G</td>\n",
       "      <td>0.233760</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>1</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>HGNC:28706</td>\n",
       "      <td>sterile alpha motif domain containing 11</td>\n",
       "      <td>616765.0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs80125161</td>\n",
       "      <td>1</td>\n",
       "      <td>902764</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>rs80125161</td>\n",
       "      <td>1:902764_C_G</td>\n",
       "      <td>0.321790</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>HGNC:25284</td>\n",
       "      <td>pleckstrin homology domain containing N1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000187583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs7524174</td>\n",
       "      <td>1</td>\n",
       "      <td>902997</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs7524174</td>\n",
       "      <td>1:902997_G_A</td>\n",
       "      <td>0.131170</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>HGNC:25284</td>\n",
       "      <td>pleckstrin homology domain containing N1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000187583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs3829740</td>\n",
       "      <td>1</td>\n",
       "      <td>909238</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>rs3829740</td>\n",
       "      <td>rs3829740</td>\n",
       "      <td>0.429340</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>HGNC:25284</td>\n",
       "      <td>pleckstrin homology domain containing N1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000187583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51394</th>\n",
       "      <td>rs9616937</td>\n",
       "      <td>22</td>\n",
       "      <td>51126128</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>rs9616937</td>\n",
       "      <td>22:51126128_T_A</td>\n",
       "      <td>0.408330</td>\n",
       "      <td>SHANK3</td>\n",
       "      <td>1</td>\n",
       "      <td>SHANK3</td>\n",
       "      <td>HGNC:14294</td>\n",
       "      <td>SH3 and multiple ankyrin repeat domains 3</td>\n",
       "      <td>606230.0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000251322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51395</th>\n",
       "      <td>rs1024374</td>\n",
       "      <td>22</td>\n",
       "      <td>51149320</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>rs1024374</td>\n",
       "      <td>rs1024374</td>\n",
       "      <td>0.461240</td>\n",
       "      <td>ACR</td>\n",
       "      <td>1</td>\n",
       "      <td>ACR</td>\n",
       "      <td>HGNC:126</td>\n",
       "      <td>acrosin</td>\n",
       "      <td>102480.0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000100312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51396</th>\n",
       "      <td>rs144480800</td>\n",
       "      <td>22</td>\n",
       "      <td>51151642</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs144480800</td>\n",
       "      <td>22:51151642_C_T</td>\n",
       "      <td>0.014888</td>\n",
       "      <td>ACR</td>\n",
       "      <td>1</td>\n",
       "      <td>ACR</td>\n",
       "      <td>HGNC:126</td>\n",
       "      <td>acrosin</td>\n",
       "      <td>102480.0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000100312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51397</th>\n",
       "      <td>rs5770994</td>\n",
       "      <td>22</td>\n",
       "      <td>51161019</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs5770994</td>\n",
       "      <td>rs5770994</td>\n",
       "      <td>0.482000</td>\n",
       "      <td>ACR</td>\n",
       "      <td>1</td>\n",
       "      <td>ACR</td>\n",
       "      <td>HGNC:126</td>\n",
       "      <td>acrosin</td>\n",
       "      <td>102480.0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000100312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51398</th>\n",
       "      <td>rs9616824</td>\n",
       "      <td>22</td>\n",
       "      <td>51180934</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs9616824</td>\n",
       "      <td>22:51180934_C_T</td>\n",
       "      <td>0.230140</td>\n",
       "      <td>ACR</td>\n",
       "      <td>1</td>\n",
       "      <td>ACR</td>\n",
       "      <td>HGNC:126</td>\n",
       "      <td>acrosin</td>\n",
       "      <td>102480.0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000100312</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>51368 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               SNP  chr       pos ref alt         rsid            varid   \n",
       "0        rs6657440    1    850780   C   T    rs6657440        rs6657440  \\\n",
       "1        rs7418179    1    858801   A   G    rs7418179     1:858801_A_G   \n",
       "2       rs80125161    1    902764   C   G   rs80125161     1:902764_C_G   \n",
       "3        rs7524174    1    902997   G   A    rs7524174     1:902997_G_A   \n",
       "4        rs3829740    1    909238   G   C    rs3829740        rs3829740   \n",
       "...            ...  ...       ...  ..  ..          ...              ...   \n",
       "51394    rs9616937   22  51126128   T   A    rs9616937  22:51126128_T_A   \n",
       "51395    rs1024374   22  51149320   C   G    rs1024374        rs1024374   \n",
       "51396  rs144480800   22  51151642   C   T  rs144480800  22:51151642_C_T   \n",
       "51397    rs5770994   22  51161019   C   T    rs5770994        rs5770994   \n",
       "51398    rs9616824   22  51180934   C   T    rs9616824  22:51180934_C_T   \n",
       "\n",
       "        maf_EUR name_1kg  tie Gene_name     hgnc_id   \n",
       "0      0.395650   SAMD11    1    SAMD11  HGNC:28706  \\\n",
       "1      0.233760   SAMD11    1    SAMD11  HGNC:28706   \n",
       "2      0.321790  PLEKHN1    1   PLEKHN1  HGNC:25284   \n",
       "3      0.131170  PLEKHN1    1   PLEKHN1  HGNC:25284   \n",
       "4      0.429340  PLEKHN1    1   PLEKHN1  HGNC:25284   \n",
       "...         ...      ...  ...       ...         ...   \n",
       "51394  0.408330   SHANK3    1    SHANK3  HGNC:14294   \n",
       "51395  0.461240      ACR    1       ACR    HGNC:126   \n",
       "51396  0.014888      ACR    1       ACR    HGNC:126   \n",
       "51397  0.482000      ACR    1       ACR    HGNC:126   \n",
       "51398  0.230140      ACR    1       ACR    HGNC:126   \n",
       "\n",
       "                                            name   omim_id  notfound   \n",
       "0       sterile alpha motif domain containing 11  616765.0         0  \\\n",
       "1       sterile alpha motif domain containing 11  616765.0         0   \n",
       "2       pleckstrin homology domain containing N1       NaN         0   \n",
       "3       pleckstrin homology domain containing N1       NaN         0   \n",
       "4       pleckstrin homology domain containing N1       NaN         0   \n",
       "...                                          ...       ...       ...   \n",
       "51394  SH3 and multiple ankyrin repeat domains 3  606230.0         0   \n",
       "51395                                    acrosin  102480.0         0   \n",
       "51396                                    acrosin  102480.0         0   \n",
       "51397                                    acrosin  102480.0         0   \n",
       "51398                                    acrosin  102480.0         0   \n",
       "\n",
       "       ensembl_gene_id  \n",
       "0      ENSG00000187634  \n",
       "1      ENSG00000187634  \n",
       "2      ENSG00000187583  \n",
       "3      ENSG00000187583  \n",
       "4      ENSG00000187583  \n",
       "...                ...  \n",
       "51394  ENSG00000251322  \n",
       "51395  ENSG00000100312  \n",
       "51396  ENSG00000100312  \n",
       "51397  ENSG00000100312  \n",
       "51398  ENSG00000100312  \n",
       "\n",
       "[51368 rows x 16 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_df_selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fa3b1769-729b-4fcc-8ddd-6e48b01ed9db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n"
     ]
    }
   ],
   "source": [
    "snplist = dict()\n",
    "for chrm in range(1,23):\n",
    "    print(chrm)\n",
    "    snplist[chrm] = variant_df_selected.query(f\"chr == {chrm}\")[\"SNP\"].tolist()\n",
    "    snp_filename = f\"{variant_dir}/snps.{chrm}.txt\"\n",
    "    with open(snp_filename, 'w') as mfile:\n",
    "        for rsid in snplist[chrm]:\n",
    "            mfile.write(f\"{rsid}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "336ffa8b-796e-4de3-84e5-15f83b7ada1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[4718,\n",
       " 4623,\n",
       " 3245,\n",
       " 2217,\n",
       " 2717,\n",
       " 3381,\n",
       " 2683,\n",
       " 2525,\n",
       " 2227,\n",
       " 2299,\n",
       " 2748,\n",
       " 2812,\n",
       " 1084,\n",
       " 1333,\n",
       " 1808,\n",
       " 2290,\n",
       " 2279,\n",
       " 938,\n",
       " 2158,\n",
       " 1861,\n",
       " 456,\n",
       " 966]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[len(x) for chrm, x in snplist.items()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12528f5f-97b6-404f-b94b-ce00c3e9da34",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
