<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.551">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Saikat Banerjee">
<meta name="dcterms.date" content="2024-01-30">

<title>NPD Network - Prefiltering data from PanUKB</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">NPD Network</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../notebooks/index.html"> 
<span class="menu-text">Notebooks</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../meetings/index.html"> 
<span class="menu-text">Meetings</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../resources/index.html"> 
<span class="menu-text">Resources</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../reference/index.html"> 
<span class="menu-text">Reference</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Contents</h2>
   
  <ul>
  <li><a href="#selection-for-matrix-completion-benchmark" id="toc-selection-for-matrix-completion-benchmark" class="nav-link active" data-scroll-target="#selection-for-matrix-completion-benchmark">Selection for matrix completion benchmark</a></li>
  <li><a href="#hierarchical-clustering-of-covariance-matrix" id="toc-hierarchical-clustering-of-covariance-matrix" class="nav-link" data-scroll-target="#hierarchical-clustering-of-covariance-matrix">Hierarchical clustering of covariance matrix</a></li>
  <li><a href="#save-the-dataframes" id="toc-save-the-dataframes" class="nav-link" data-scroll-target="#save-the-dataframes">Save the dataframes</a></li>
  <li><a href="#rpca---ialm" id="toc-rpca---ialm" class="nav-link" data-scroll-target="#rpca---ialm">RPCA - IALM</a></li>
  <li><a href="#nnm" id="toc-nnm" class="nav-link" data-scroll-target="#nnm">NNM</a></li>
  <li><a href="#nnm-sparse" id="toc-nnm-sparse" class="nav-link" data-scroll-target="#nnm-sparse">NNM-Sparse</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.dev/banskt/npd-notes/blob/main/notebooks/ukbb/2024-01-20-panukb-data-filtering-for-masked-input.ipynb" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/banskt/npd-notes/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Prefiltering data from PanUKB</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Saikat Banerjee </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 30, 2024</p>
    </div>
  </div>
  
    
  </div>
  
<div>
  <div class="abstract">
    <div class="block-title">Abstract</div>
    We read the data from PanUKB and prefilter based on heritability and effect size.
  </div>
</div>


</header>


<div id="6c819d42-dd9d-4b57-b466-399d674567d7" class="cell" data-execution_count="1">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pickle</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy <span class="im">import</span> stats <span class="im">as</span> sc_stats</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pymir <span class="im">import</span> mpl_stylesheet</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pymir <span class="im">import</span> mpl_utils</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>mpl_stylesheet.banskt_presentation(splinecolor <span class="op">=</span> <span class="st">'black'</span>, dpi <span class="op">=</span> <span class="dv">120</span>, colors <span class="op">=</span> <span class="st">'kelly'</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib.gridspec <span class="im">import</span> GridSpec</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> nnwmf.optimize <span class="im">import</span> IALM</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> nnwmf.optimize <span class="im">import</span> FrankWolfe, FrankWolfe_CV</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> nnwmf.utils <span class="im">import</span> model_errors <span class="im">as</span> merr</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>sys.path.append(<span class="st">"../utils/"</span>)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> histogram <span class="im">as</span> mpy_histogram</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> simulate <span class="im">as</span> mpy_simulate</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> plot_functions <span class="im">as</span> mpy_plotfn</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="6baa9bbc-bfbd-4a4e-b556-c9f4059ea1c8" class="cell" data-execution_count="2">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>data_dir <span class="op">=</span> <span class="st">"/gpfs/commons/home/sbanerjee/work/npd/PanUKB/data"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>result_dir <span class="op">=</span> <span class="st">"/gpfs/commons/home/sbanerjee/work/npd/PanUKB/nnwmf"</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>zscore_filename <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>data_dir<span class="sc">}</span><span class="ss">/GWAS_Zscore.tsv"</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>trait_filename <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>data_dir<span class="sc">}</span><span class="ss">/trait_manifest_TableS6_no_readme.tsv"</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>variant_filename <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>data_dir<span class="sc">}</span><span class="ss">/allvar.pruned.closesttss.hugo"</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>zscore_df <span class="op">=</span> pd.read_csv(zscore_filename, sep <span class="op">=</span> <span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>trait_df <span class="op">=</span> pd.read_csv(trait_filename, sep <span class="op">=</span> <span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>variant_df <span class="op">=</span> pd.read_csv(variant_filename, sep <span class="op">=</span> <span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="co"># remove extra columns from trait_df</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>colnames <span class="op">=</span> trait_df.columns.tolist()</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>colnames[<span class="dv">0</span>] <span class="op">=</span> <span class="st">"zindex"</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>trait_df.columns <span class="op">=</span> colnames</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>trait_df_mod <span class="op">=</span> trait_df.drop(labels <span class="op">=</span> [<span class="st">'coding'</span>, <span class="st">'modifier'</span>, <span class="st">'coding_description'</span>, <span class="st">'filename'</span>, <span class="st">'aws_link'</span>], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="co">#trait_df_mod</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="b6b1d24e-147d-4915-80d4-2b9491beaa0f" class="cell" data-execution_count="3">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>trait_df_mod</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="3">
<div>
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">zindex</th>
<th data-quarto-table-cell-role="th">trait_type</th>
<th data-quarto-table-cell-role="th">phenocode</th>
<th data-quarto-table-cell-role="th">pheno_sex</th>
<th data-quarto-table-cell-role="th">description</th>
<th data-quarto-table-cell-role="th">description_more</th>
<th data-quarto-table-cell-role="th">category</th>
<th data-quarto-table-cell-role="th">BIN_QT</th>
<th data-quarto-table-cell-role="th">n_cases_EUR</th>
<th data-quarto-table-cell-role="th">n_controls_EUR</th>
<th data-quarto-table-cell-role="th">N</th>
<th data-quarto-table-cell-role="th">Neff</th>
<th data-quarto-table-cell-role="th">estimates.final.h2_observed</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>1</td>
<td>icd10</td>
<td>A04</td>
<td>both_sexes</td>
<td>A04 Other bacterial intestinal infections</td>
<td>truncated: true</td>
<td>Chapter I Certain infectious and parasitic dis...</td>
<td>BIN</td>
<td>3088</td>
<td>417443.0</td>
<td>420531</td>
<td>6130.649032</td>
<td>0.0033</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>2</td>
<td>icd10</td>
<td>A08</td>
<td>both_sexes</td>
<td>A08 Viral and other specified intestinal infec...</td>
<td>truncated: true</td>
<td>Chapter I Certain infectious and parasitic dis...</td>
<td>BIN</td>
<td>1107</td>
<td>419424.0</td>
<td>420531</td>
<td>2208.171897</td>
<td>0.0001</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>3</td>
<td>icd10</td>
<td>A09</td>
<td>both_sexes</td>
<td>A09 Diarrhoea and gastro-enteritis of presumed...</td>
<td>truncated: true</td>
<td>Chapter I Certain infectious and parasitic dis...</td>
<td>BIN</td>
<td>9029</td>
<td>411502.0</td>
<td>420531</td>
<td>17670.286180</td>
<td>0.0035</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>4</td>
<td>icd10</td>
<td>A41</td>
<td>both_sexes</td>
<td>A41 Other septicaemia</td>
<td>truncated: true</td>
<td>Chapter I Certain infectious and parasitic dis...</td>
<td>BIN</td>
<td>5512</td>
<td>415019.0</td>
<td>420531</td>
<td>10879.505810</td>
<td>0.0011</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>5</td>
<td>icd10</td>
<td>B34</td>
<td>both_sexes</td>
<td>B34 Viral infection of unspecified site</td>
<td>truncated: true</td>
<td>Chapter I Certain infectious and parasitic dis...</td>
<td>BIN</td>
<td>2129</td>
<td>418402.0</td>
<td>420531</td>
<td>4236.443249</td>
<td>0.0003</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2478</td>
<td>2479</td>
<td>continuous</td>
<td>Smoking</td>
<td>both_sexes</td>
<td>Smoking status, ever vs never</td>
<td>Ever (previous + current smoker) vs never base...</td>
<td>NaN</td>
<td>QT</td>
<td>418817</td>
<td>NaN</td>
<td>418817</td>
<td>418817.000000</td>
<td>0.1100</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2479</td>
<td>2480</td>
<td>continuous</td>
<td>eGFR</td>
<td>both_sexes</td>
<td>Estimated glomerular filtration rate, serum cr...</td>
<td>eGFR based on serum creatinine (30700) using t...</td>
<td>NaN</td>
<td>QT</td>
<td>401867</td>
<td>NaN</td>
<td>401867</td>
<td>401867.000000</td>
<td>0.2070</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2480</td>
<td>2481</td>
<td>continuous</td>
<td>eGFRcreacys</td>
<td>both_sexes</td>
<td>Estimated glomerular filtration rate, cystain C</td>
<td>eGFR based on cystain C (30720) using the CKD-...</td>
<td>NaN</td>
<td>QT</td>
<td>401570</td>
<td>NaN</td>
<td>401570</td>
<td>401570.000000</td>
<td>0.2380</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2481</td>
<td>2482</td>
<td>continuous</td>
<td>eGFRcys</td>
<td>both_sexes</td>
<td>Estimated glomerular filtration rate, serum cr...</td>
<td>eGFR based on serum creatinine (30700) and cys...</td>
<td>NaN</td>
<td>QT</td>
<td>402031</td>
<td>NaN</td>
<td>402031</td>
<td>402031.000000</td>
<td>0.2240</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2482</td>
<td>2483</td>
<td>continuous</td>
<td>whr</td>
<td>both_sexes</td>
<td>pheno 48 / pheno 49</td>
<td>NaN</td>
<td>NaN</td>
<td>QT</td>
<td>420531</td>
<td>NaN</td>
<td>420531</td>
<td>420531.000000</td>
<td>0.1740</td>
</tr>
</tbody>
</table>

<p>2483 rows × 13 columns</p>
</div>
</div>
</div>
</div>
<div id="8738e50d-aeca-42cc-92c2-d3659e63a2c8" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>zscore_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="4">
<div>
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">rsid</th>
<th data-quarto-table-cell-role="th">z1</th>
<th data-quarto-table-cell-role="th">z2</th>
<th data-quarto-table-cell-role="th">z3</th>
<th data-quarto-table-cell-role="th">z4</th>
<th data-quarto-table-cell-role="th">z5</th>
<th data-quarto-table-cell-role="th">z6</th>
<th data-quarto-table-cell-role="th">z7</th>
<th data-quarto-table-cell-role="th">z8</th>
<th data-quarto-table-cell-role="th">z9</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">z2474</th>
<th data-quarto-table-cell-role="th">z2475</th>
<th data-quarto-table-cell-role="th">z2476</th>
<th data-quarto-table-cell-role="th">z2477</th>
<th data-quarto-table-cell-role="th">z2478</th>
<th data-quarto-table-cell-role="th">z2479</th>
<th data-quarto-table-cell-role="th">z2480</th>
<th data-quarto-table-cell-role="th">z2481</th>
<th data-quarto-table-cell-role="th">z2482</th>
<th data-quarto-table-cell-role="th">z2483</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>rs6657440</td>
<td>-0.903532</td>
<td>0.561842</td>
<td>0.711068</td>
<td>-0.109174</td>
<td>0.223668</td>
<td>-1.728199</td>
<td>0.374988</td>
<td>-0.265971</td>
<td>-2.823282</td>
<td>...</td>
<td>1.521092</td>
<td>0.612532</td>
<td>1.405428</td>
<td>0.018029</td>
<td>0.895337</td>
<td>-0.008761</td>
<td>-2.069432</td>
<td>-4.292948</td>
<td>-4.701711</td>
<td>2.952899</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>rs7418179</td>
<td>0.398166</td>
<td>1.163539</td>
<td>0.512118</td>
<td>0.144794</td>
<td>-1.313903</td>
<td>-1.547410</td>
<td>0.450270</td>
<td>0.560324</td>
<td>-1.502268</td>
<td>...</td>
<td>-0.296537</td>
<td>-0.734266</td>
<td>-0.093081</td>
<td>0.412077</td>
<td>1.961159</td>
<td>0.716049</td>
<td>-2.171984</td>
<td>-5.314085</td>
<td>-6.612137</td>
<td>3.817518</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>rs80125161</td>
<td>-1.739115</td>
<td>-0.172328</td>
<td>0.349145</td>
<td>-0.329335</td>
<td>-0.870640</td>
<td>-1.004155</td>
<td>1.128148</td>
<td>0.151244</td>
<td>-1.816075</td>
<td>...</td>
<td>2.222433</td>
<td>1.092969</td>
<td>2.328233</td>
<td>1.160767</td>
<td>0.909524</td>
<td>-1.467249</td>
<td>-0.135785</td>
<td>-2.187241</td>
<td>-3.223529</td>
<td>4.508578</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>rs7524174</td>
<td>-0.884478</td>
<td>-1.762000</td>
<td>1.312823</td>
<td>-0.550764</td>
<td>2.132540</td>
<td>0.519828</td>
<td>0.834194</td>
<td>0.699441</td>
<td>-0.885281</td>
<td>...</td>
<td>3.356354</td>
<td>1.990588</td>
<td>3.092179</td>
<td>-0.133810</td>
<td>-0.072845</td>
<td>-1.376310</td>
<td>1.317044</td>
<td>0.913491</td>
<td>0.535188</td>
<td>2.245657</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>rs3829740</td>
<td>-1.469931</td>
<td>-0.519628</td>
<td>-0.281605</td>
<td>-0.267729</td>
<td>-1.060167</td>
<td>0.058116</td>
<td>-0.638319</td>
<td>-0.589767</td>
<td>0.228514</td>
<td>...</td>
<td>-0.320075</td>
<td>-0.128047</td>
<td>-0.524757</td>
<td>-0.232900</td>
<td>-1.051020</td>
<td>-0.483644</td>
<td>2.026508</td>
<td>4.400092</td>
<td>5.407316</td>
<td>1.125536</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">51394</td>
<td>rs9616937</td>
<td>-0.211947</td>
<td>1.371231</td>
<td>-1.800776</td>
<td>0.609980</td>
<td>-0.619822</td>
<td>0.947269</td>
<td>-1.166021</td>
<td>0.478601</td>
<td>-0.359714</td>
<td>...</td>
<td>0.714167</td>
<td>0.354347</td>
<td>0.611158</td>
<td>-0.354725</td>
<td>1.073043</td>
<td>-0.831737</td>
<td>0.870924</td>
<td>1.432076</td>
<td>2.228501</td>
<td>0.536104</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">51395</td>
<td>rs1024374</td>
<td>0.027097</td>
<td>-1.817082</td>
<td>0.530216</td>
<td>0.813498</td>
<td>-0.076514</td>
<td>0.784427</td>
<td>1.411160</td>
<td>-1.111740</td>
<td>-0.224438</td>
<td>...</td>
<td>1.107098</td>
<td>1.482684</td>
<td>1.512723</td>
<td>0.322355</td>
<td>-0.374603</td>
<td>1.320194</td>
<td>-0.700092</td>
<td>-1.395039</td>
<td>-2.270186</td>
<td>0.360025</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">51396</td>
<td>rs144480800</td>
<td>0.545682</td>
<td>0.391830</td>
<td>0.520505</td>
<td>-1.280976</td>
<td>0.453876</td>
<td>-1.388940</td>
<td>0.025094</td>
<td>0.737788</td>
<td>1.178641</td>
<td>...</td>
<td>-0.562063</td>
<td>-1.148515</td>
<td>-0.994185</td>
<td>-0.268232</td>
<td>-0.069619</td>
<td>0.013256</td>
<td>-0.777667</td>
<td>-1.544760</td>
<td>-1.406344</td>
<td>2.205817</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">51397</td>
<td>rs5770994</td>
<td>1.441851</td>
<td>1.152368</td>
<td>-1.500000</td>
<td>-0.468137</td>
<td>-0.444156</td>
<td>-0.780139</td>
<td>-0.853550</td>
<td>-0.316097</td>
<td>0.311219</td>
<td>...</td>
<td>-1.185702</td>
<td>-0.624073</td>
<td>-0.859522</td>
<td>0.549669</td>
<td>1.809912</td>
<td>0.268733</td>
<td>0.947441</td>
<td>1.533302</td>
<td>1.658537</td>
<td>2.218653</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">51398</td>
<td>rs9616824</td>
<td>0.669372</td>
<td>-0.308184</td>
<td>-0.811311</td>
<td>-0.154224</td>
<td>1.476823</td>
<td>0.065167</td>
<td>-1.030063</td>
<td>-0.215723</td>
<td>-0.631385</td>
<td>...</td>
<td>0.308486</td>
<td>0.612102</td>
<td>0.492977</td>
<td>-0.621278</td>
<td>-2.657386</td>
<td>-0.732402</td>
<td>-0.106474</td>
<td>-0.029006</td>
<td>0.070647</td>
<td>-0.805253</td>
</tr>
</tbody>
</table>

<p>51399 rows × 2484 columns</p>
</div>
</div>
</div>
</div>
<section id="selection-for-matrix-completion-benchmark" class="level1">
<h1>Selection for matrix completion benchmark</h1>
<div id="b573272b-e635-4fb0-981b-0ee513c68b1f" class="cell" data-execution_count="6">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>h2_cut <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>pval_cut <span class="op">=</span> <span class="fl">5e-8</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>trait_df_selected <span class="op">=</span> trait_df_mod[trait_df_mod[<span class="st">'estimates.final.h2_observed'</span>] <span class="op">&gt;</span> h2_cut]</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>zidx_selected <span class="op">=</span> trait_df_selected[<span class="st">'zindex'</span>].values</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>colnames_selected <span class="op">=</span> [<span class="st">'rsid'</span>] <span class="op">+</span> [<span class="ss">f'z</span><span class="sc">{</span>x<span class="sc">}</span><span class="ss">'</span> <span class="cf">for</span> x <span class="kw">in</span> zidx_selected]</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>zscore_df_selected <span class="op">=</span> zscore_df[colnames_selected]</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>zcut <span class="op">=</span> sc_stats.norm.ppf(<span class="dv">1</span> <span class="op">-</span> (pval_cut <span class="op">/</span> <span class="dv">2</span>))</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>rsid_idx_selected <span class="op">=</span> []</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> x <span class="kw">in</span> zidx_selected:</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>    rsid_idx_selected <span class="op">+=</span> zscore_df_selected[zscore_df_selected[<span class="ss">f'z</span><span class="sc">{</span>x<span class="sc">}</span><span class="ss">'</span>].<span class="bu">abs</span>() <span class="op">&gt;</span> zcut].index.to_list()</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>rsid_idx_selected <span class="op">=</span> <span class="bu">sorted</span>(<span class="bu">set</span>(rsid_idx_selected))</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>zscore_df_selected <span class="op">=</span> zscore_df_selected.iloc[rsid_idx_selected, :].drop(labels <span class="op">=</span> [<span class="st">'rsid'</span>], axis <span class="op">=</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="e7402b24-4b91-4103-8ddd-67fcd40b446d" class="cell" data-execution_count="7">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>X <span class="op">=</span> np.array(zscore_df_selected.values.T)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>X_cent <span class="op">=</span> X <span class="op">-</span> np.mean(X, axis <span class="op">=</span> <span class="dv">0</span>, keepdims <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"We have </span><span class="sc">{</span>X_cent<span class="sc">.</span>shape[<span class="dv">0</span>]<span class="sc">}</span><span class="ss"> samples (phenotypes) and </span><span class="sc">{</span>X_cent<span class="sc">.</span>shape[<span class="dv">1</span>]<span class="sc">}</span><span class="ss"> features (variants)"</span>)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Fraction of Nan entries: </span><span class="sc">{</span>np<span class="sc">.</span><span class="bu">sum</span>(np.isnan(X)) <span class="op">/</span> np<span class="sc">.</span>prod(X_cent.shape)<span class="sc">:.3f}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>We have 216 samples (phenotypes) and 48212 features (variants)
Fraction of Nan entries: 0.000</code></pre>
</div>
</div>
<div id="581e098e-b932-438d-8099-d5f224d20af5" class="cell" data-execution_count="8">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy.stats <span class="im">as</span> sc_stats</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_density(x, data):</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    density <span class="op">=</span> sc_stats.gaussian_kde(data)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> density.pdf(x)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_bins(data, nbin, xmin, xmax):</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>    xdelta <span class="op">=</span> (np.<span class="bu">max</span>(data) <span class="op">-</span> np.<span class="bu">min</span>(data)) <span class="op">/</span> <span class="dv">10</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> xmin: xmin <span class="op">=</span> np.<span class="bu">min</span>(data) <span class="op">-</span> xdelta</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> xmax: xmax <span class="op">=</span> np.<span class="bu">max</span>(data) <span class="op">+</span> xdelta</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>    bins <span class="op">=</span> np.linspace(xmin, xmax, nbin)</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>    xbin <span class="op">=</span> [(bins[i] <span class="op">+</span> bins[i<span class="op">+</span><span class="dv">1</span>]) <span class="op">/</span> <span class="dv">2</span> <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(bins.shape[<span class="dv">0</span>] <span class="op">-</span> <span class="dv">1</span>)] <span class="co"># centers of the bins</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> xmin, xmax, bins, xbin</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure()</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>ax1 <span class="op">=</span> fig.add_subplot(<span class="dv">111</span>)</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> n <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">100</span>):</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>    scores_cent <span class="op">=</span> X_cent[n, :]</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>    xmin, xmax, bins, xbin <span class="op">=</span> get_bins(scores_cent, <span class="dv">100</span>, <span class="op">-</span><span class="dv">10</span>, <span class="dv">10</span>)</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>    curve <span class="op">=</span> get_density(xbin, scores_cent)</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>    ax1.plot(xbin, curve)</span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2024-01-20-panukb-data-filtering-for-masked-input_files/figure-html/cell-8-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="hierarchical-clustering-of-covariance-matrix" class="level1">
<h1>Hierarchical clustering of covariance matrix</h1>
<div id="37b49d9c-1c3c-43d9-aedc-a03308f798f6" class="cell" data-execution_count="9">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.cluster <span class="im">import</span> hierarchy <span class="im">as</span> sp_hierarchy</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>sample_distance_matrix <span class="op">=</span> np.cov(X_cent)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>o1 <span class="op">=</span> sp_hierarchy.leaves_list(sp_hierarchy.linkage(sample_distance_matrix, method <span class="op">=</span> <span class="st">'centroid'</span>))</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure(figsize <span class="op">=</span> (<span class="dv">8</span>, <span class="dv">8</span>))</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>ax1 <span class="op">=</span> fig.add_subplot(<span class="dv">111</span>)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>                                            </span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="co">#mpy_plotfn.plot_covariance_heatmap(ax1, Z * np.sqrt(p) / np.sqrt(nsample))</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>mpy_plotfn.plot_covariance_heatmap(ax1, X_cent[o1, :])</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2024-01-20-panukb-data-filtering-for-masked-input_files/figure-html/cell-9-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="save-the-dataframes" class="level1">
<h1>Save the dataframes</h1>
<div id="35932908-cab1-4cbc-863e-c6bdc1f2925c" class="cell" data-execution_count="54">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>zscore_df_selected_filename <span class="op">=</span> os.path.join(data_dir, <span class="ss">f"modselect/zscore_h2</span><span class="sc">{</span>h2_cut<span class="sc">}</span><span class="ss">_pval</span><span class="sc">{</span>pval_cut<span class="sc">}</span><span class="ss">.pkl"</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>trait_df_selected_filename  <span class="op">=</span> os.path.join(data_dir, <span class="ss">f"modselect/traits_h2</span><span class="sc">{</span>h2_cut<span class="sc">}</span><span class="ss">.pkl"</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># zscore_df_selected.to_pickle(zscore_df_selected_filename)</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co"># trait_df_selected.to_pickle(trait_df_selected_filename)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="e5478a62-27c9-4a11-87a3-784384507ed2" class="cell" data-execution_count="10">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>zscore_df_selected</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="10">
<div>
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">z1415</th>
<th data-quarto-table-cell-role="th">z1678</th>
<th data-quarto-table-cell-role="th">z1679</th>
<th data-quarto-table-cell-role="th">z1680</th>
<th data-quarto-table-cell-role="th">z1681</th>
<th data-quarto-table-cell-role="th">z1682</th>
<th data-quarto-table-cell-role="th">z1683</th>
<th data-quarto-table-cell-role="th">z1685</th>
<th data-quarto-table-cell-role="th">z1686</th>
<th data-quarto-table-cell-role="th">z1687</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">z2473</th>
<th data-quarto-table-cell-role="th">z2474</th>
<th data-quarto-table-cell-role="th">z2475</th>
<th data-quarto-table-cell-role="th">z2476</th>
<th data-quarto-table-cell-role="th">z2477</th>
<th data-quarto-table-cell-role="th">z2479</th>
<th data-quarto-table-cell-role="th">z2480</th>
<th data-quarto-table-cell-role="th">z2481</th>
<th data-quarto-table-cell-role="th">z2482</th>
<th data-quarto-table-cell-role="th">z2483</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>3.591973</td>
<td>-0.038545</td>
<td>4.356103</td>
<td>-0.030974</td>
<td>-2.625193</td>
<td>-0.082182</td>
<td>-1.513514</td>
<td>2.927977</td>
<td>1.145769</td>
<td>-0.924589</td>
<td>...</td>
<td>-0.964337</td>
<td>1.521092</td>
<td>0.612532</td>
<td>1.405428</td>
<td>0.018029</td>
<td>-0.008761</td>
<td>-2.069432</td>
<td>-4.292948</td>
<td>-4.701711</td>
<td>2.952899</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>2.690253</td>
<td>2.060959</td>
<td>5.074098</td>
<td>2.131896</td>
<td>-2.200532</td>
<td>0.203585</td>
<td>0.053289</td>
<td>2.081926</td>
<td>1.606040</td>
<td>-0.265317</td>
<td>...</td>
<td>-0.158982</td>
<td>-0.296537</td>
<td>-0.734266</td>
<td>-0.093081</td>
<td>0.412077</td>
<td>0.716049</td>
<td>-2.171984</td>
<td>-5.314085</td>
<td>-6.612137</td>
<td>3.817518</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>4.672697</td>
<td>1.966395</td>
<td>6.363325</td>
<td>1.017074</td>
<td>-3.506709</td>
<td>-0.985765</td>
<td>-0.432902</td>
<td>2.834796</td>
<td>3.055340</td>
<td>-2.151471</td>
<td>...</td>
<td>0.716255</td>
<td>2.222433</td>
<td>1.092969</td>
<td>2.328233</td>
<td>1.160767</td>
<td>-1.467249</td>
<td>-0.135785</td>
<td>-2.187241</td>
<td>-3.223529</td>
<td>4.508578</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>3.628734</td>
<td>0.377612</td>
<td>1.851399</td>
<td>-1.249221</td>
<td>-1.881565</td>
<td>-0.479096</td>
<td>-1.015920</td>
<td>0.498607</td>
<td>-0.294269</td>
<td>-1.046150</td>
<td>...</td>
<td>-1.404276</td>
<td>3.356354</td>
<td>1.990588</td>
<td>3.092179</td>
<td>-0.133810</td>
<td>-1.376310</td>
<td>1.317044</td>
<td>0.913491</td>
<td>0.535188</td>
<td>2.245657</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>-0.858911</td>
<td>-0.623239</td>
<td>-1.686601</td>
<td>-3.003435</td>
<td>0.532031</td>
<td>-0.341693</td>
<td>-3.777505</td>
<td>0.254496</td>
<td>0.226082</td>
<td>-0.322407</td>
<td>...</td>
<td>-0.536635</td>
<td>-0.320075</td>
<td>-0.128047</td>
<td>-0.524757</td>
<td>-0.232900</td>
<td>-0.483644</td>
<td>2.026508</td>
<td>4.400092</td>
<td>5.407316</td>
<td>1.125536</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">51394</td>
<td>1.342127</td>
<td>1.797208</td>
<td>-0.979370</td>
<td>-1.867291</td>
<td>0.037906</td>
<td>-1.162637</td>
<td>-3.256483</td>
<td>2.176569</td>
<td>-1.732998</td>
<td>-0.328859</td>
<td>...</td>
<td>0.482023</td>
<td>0.714167</td>
<td>0.354347</td>
<td>0.611158</td>
<td>-0.354725</td>
<td>-0.831737</td>
<td>0.870924</td>
<td>1.432076</td>
<td>2.228501</td>
<td>0.536104</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">51395</td>
<td>-1.244520</td>
<td>-1.527636</td>
<td>1.849772</td>
<td>2.144946</td>
<td>-0.031816</td>
<td>2.274951</td>
<td>2.456148</td>
<td>-2.560491</td>
<td>-0.212461</td>
<td>0.982478</td>
<td>...</td>
<td>-0.110369</td>
<td>1.107098</td>
<td>1.482684</td>
<td>1.512723</td>
<td>0.322355</td>
<td>1.320194</td>
<td>-0.700092</td>
<td>-1.395039</td>
<td>-2.270186</td>
<td>0.360025</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">51396</td>
<td>-0.253862</td>
<td>0.812630</td>
<td>-0.711050</td>
<td>1.179577</td>
<td>-2.862653</td>
<td>-2.013295</td>
<td>2.833169</td>
<td>0.153290</td>
<td>2.138158</td>
<td>-2.653582</td>
<td>...</td>
<td>-0.088946</td>
<td>-0.562063</td>
<td>-1.148515</td>
<td>-0.994185</td>
<td>-0.268232</td>
<td>0.013256</td>
<td>-0.777667</td>
<td>-1.544760</td>
<td>-1.406344</td>
<td>2.205817</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">51397</td>
<td>-0.915597</td>
<td>1.668377</td>
<td>-0.823597</td>
<td>0.163447</td>
<td>-0.061726</td>
<td>0.319698</td>
<td>0.922030</td>
<td>2.559199</td>
<td>0.215736</td>
<td>0.468609</td>
<td>...</td>
<td>0.098326</td>
<td>-1.185702</td>
<td>-0.624073</td>
<td>-0.859522</td>
<td>0.549669</td>
<td>0.268733</td>
<td>0.947441</td>
<td>1.533302</td>
<td>1.658537</td>
<td>2.218653</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">51398</td>
<td>1.047521</td>
<td>-0.022502</td>
<td>2.343549</td>
<td>-0.933287</td>
<td>1.285805</td>
<td>-1.105163</td>
<td>0.234492</td>
<td>-0.452837</td>
<td>-0.142916</td>
<td>-0.777741</td>
<td>...</td>
<td>0.713318</td>
<td>0.308486</td>
<td>0.612102</td>
<td>0.492977</td>
<td>-0.621278</td>
<td>-0.732402</td>
<td>-0.106474</td>
<td>-0.029006</td>
<td>0.070647</td>
<td>-0.805253</td>
</tr>
</tbody>
</table>

<p>48212 rows × 216 columns</p>
</div>
</div>
</div>
</div>
<p><strong>MHC region</strong> (hg38)</p>
<p>Chromosome 6: 28510120 - 33480577</p>
<div id="d2544299-82d7-4ed8-ab01-c6b347a4477a" class="cell" data-execution_count="34">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>variant_df.query(<span class="st">'chr == 6 and 28510120 &lt; pos &lt; 33480577'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="34">
<div>
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">SNP</th>
<th data-quarto-table-cell-role="th">chr</th>
<th data-quarto-table-cell-role="th">pos</th>
<th data-quarto-table-cell-role="th">ref</th>
<th data-quarto-table-cell-role="th">alt</th>
<th data-quarto-table-cell-role="th">rsid</th>
<th data-quarto-table-cell-role="th">varid</th>
<th data-quarto-table-cell-role="th">maf_EUR</th>
<th data-quarto-table-cell-role="th">name_1kg</th>
<th data-quarto-table-cell-role="th">tie</th>
<th data-quarto-table-cell-role="th">Gene_name</th>
<th data-quarto-table-cell-role="th">hgnc_id</th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">omim_id</th>
<th data-quarto-table-cell-role="th">notfound</th>
<th data-quarto-table-cell-role="th">ensembl_gene_id</th>
</tr>
</thead>
<tbody>
</tbody>
</table>

</div>
</div>
</div>
</div>
<section id="generate-a-masked-input" class="level2">
<h2 class="anchored" data-anchor-id="generate-a-masked-input">Generate a masked input</h2>
<div id="4218b877-148c-49b2-bbd9-23c2b28b9a50" class="cell" data-execution_count="28">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> generate_masked_input(Y, mask):</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    Ymiss_nan <span class="op">=</span> Y.copy()</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    Ymiss_nan[mask] <span class="op">=</span> np.nan</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    Ymiss_nan_cent <span class="op">=</span> Ymiss_nan <span class="op">-</span> np.nanmean(Ymiss_nan, axis <span class="op">=</span> <span class="dv">0</span>, keepdims <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    Ymiss_nan_cent[mask] <span class="op">=</span> <span class="fl">0.0</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> Ymiss_nan_cent</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> generate_mask(n, p, ratio):</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    mask <span class="op">=</span> np.ones(n <span class="op">*</span> p)</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>    nzero <span class="op">=</span> <span class="bu">int</span>(ratio <span class="op">*</span> n <span class="op">*</span> p)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>    mask[:nzero] <span class="op">=</span> <span class="fl">0.0</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>    np.random.shuffle(mask)</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> mask.reshape(n,p) <span class="op">==</span> <span class="fl">0.</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>Z_mask <span class="op">=</span> generate_mask(X.shape[<span class="dv">0</span>], X.shape[<span class="dv">1</span>], <span class="fl">0.2</span>)</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>Z_cent <span class="op">=</span> generate_masked_input(X_cent, Z_mask)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="2ba1aebf-49f8-4c26-b58a-15499291c198" class="cell" data-execution_count="29">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> root_mean_squared_error(original, recovered, mask <span class="op">=</span> <span class="va">None</span>):</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> mask <span class="kw">is</span> <span class="va">None</span>: mask <span class="op">=</span> np.ones_like(original)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    n <span class="op">=</span> np.<span class="bu">sum</span>(mask)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    mse <span class="op">=</span> np.<span class="bu">sum</span>(np.square((original <span class="op">-</span> recovered) <span class="op">*</span> mask)) <span class="op">/</span> n</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> np.sqrt(mse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="ec4cfe8d-d7a9-40ce-ab7e-a7345ca5f049" class="cell" data-execution_count="30">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>root_mean_squared_error(X_cent, Z_cent, mask <span class="op">=</span> Z_mask)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="30">
<pre><code>1.99756722409535</code></pre>
</div>
</div>
<div id="ef269314-194c-432e-b232-9ecfe2956ad2" class="cell" data-execution_count="31">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>oracle_pred <span class="op">=</span> X_cent.copy()</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>oracle_pred[<span class="op">~</span>Z_mask] <span class="op">=</span> <span class="fl">0.0</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>root_mean_squared_error(X_cent, oracle_pred, mask <span class="op">=</span> Z_mask)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="31">
<pre><code>0.0</code></pre>
</div>
</div>
</section>
</section>
<section id="rpca---ialm" class="level1">
<h1>RPCA - IALM</h1>
<div id="faa18eed-e236-4bea-aabc-4509fc359583" class="cell" data-execution_count="34">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>rpca_model <span class="op">=</span> IALM(max_iter <span class="op">=</span> <span class="dv">10000</span>, mu_update_method<span class="op">=</span><span class="st">'admm'</span>, show_progress <span class="op">=</span> <span class="va">True</span>, print_skip <span class="op">=</span> <span class="dv">100</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>rpca_model.fit(Z_cent, mask <span class="op">=</span> Z_mask)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>2024-02-23 13:13:17,128 | nnwmf.optimize.inexact_alm               | DEBUG   | Fit RPCA using IALM (mu update admm, lamba = 0.0046)
2024-02-23 13:13:20,398 | nnwmf.optimize.inexact_alm               | INFO    | Iteration 0. Primal residual 0.904981. Dual residual 0.000129418
2024-02-23 13:16:14,617 | nnwmf.optimize.inexact_alm               | INFO    | Iteration 100. Primal residual 5.08002e-06. Dual residual 4.58555e-06
2024-02-23 13:19:08,475 | nnwmf.optimize.inexact_alm               | INFO    | Iteration 200. Primal residual 9.00645e-07. Dual residual 9.55942e-07
2024-02-23 13:22:02,621 | nnwmf.optimize.inexact_alm               | INFO    | Iteration 300. Primal residual 2.24024e-07. Dual residual 2.41075e-07</code></pre>
</div>
</div>
<div id="f80ff1cc-e49c-46f2-bbeb-b3370b2ab6e6" class="cell" data-execution_count="37">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="st">"Nuclear Norms"</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Original matrix: </span><span class="sc">{</span>np<span class="sc">.</span>linalg<span class="sc">.</span>norm(X, <span class="st">'nuc'</span>)<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Original matrix centered: </span><span class="sc">{</span>np<span class="sc">.</span>linalg<span class="sc">.</span>norm(X_cent, <span class="st">'nuc'</span>)<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Masked matrix: </span><span class="sc">{</span>np<span class="sc">.</span>linalg<span class="sc">.</span>norm(Z_cent, <span class="st">'nuc'</span>)<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Recovered matrix: </span><span class="sc">{</span>np<span class="sc">.</span>linalg<span class="sc">.</span>norm(rpca_model.L_, <span class="st">'nuc'</span>)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Nuclear Norms
Original matrix: 56014.783744890694
Original matrix centered: 55147.50855870614
Masked matrix: 64158.87521360069
Recovered matrix: 18790.550737035624</code></pre>
</div>
</div>
<div id="8c06586e-7db6-4312-92b2-5b9984d77f80" class="cell" data-execution_count="38">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>root_mean_squared_error(X_cent, rpca_model.L_, mask <span class="op">=</span> Z_mask)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="38">
<pre><code>1.9975672115686012</code></pre>
</div>
</div>
</section>
<section id="nnm" class="level1">
<h1>NNM</h1>
<div id="c14e485f-5983-41a7-bf45-d69391a9c586" class="cell" data-execution_count="20">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>rank_opt <span class="op">=</span> <span class="fl">4096.</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>nnm_model <span class="op">=</span> FrankWolfe(model <span class="op">=</span> <span class="st">'nnm'</span>, max_iter <span class="op">=</span> <span class="dv">10000</span>, svd_max_iter <span class="op">=</span> <span class="dv">50</span>, show_progress <span class="op">=</span> <span class="va">True</span>, print_skip <span class="op">=</span> <span class="dv">100</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>nnm_model.fit(Z_cent, rank_opt, mask <span class="op">=</span> Z_mask)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>2024-02-13 10:41:55,467 | nnwmf.optimize.frankwolfe                | INFO    | Iteration 0. Step size 0.429. Duality Gap 7.1954e+06
2024-02-13 10:47:13,083 | nnwmf.optimize.frankwolfe                | INFO    | Iteration 100. Step size 0.008. Duality Gap 137328
2024-02-13 10:51:56,111 | nnwmf.optimize.frankwolfe                | INFO    | Iteration 200. Step size 0.003. Duality Gap 57230.8
2024-02-13 10:56:39,601 | nnwmf.optimize.frankwolfe                | INFO    | Iteration 300. Step size 0.003. Duality Gap 36125.2</code></pre>
</div>
</div>
<div id="11d87813-946b-48ff-b053-fcb315d826b2" class="cell" data-execution_count="28">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>root_mean_squared_error(X_cent, nnm_model.X, mask <span class="op">=</span> Z_mask)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="28">
<pre><code>1.8754185985095613</code></pre>
</div>
</div>
<div id="2b14bbe2-048e-47d6-81cb-d6dce90e7bfd" class="cell" data-execution_count="37">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>nseq  <span class="op">=</span> <span class="bu">int</span>(np.floor(np.log2(rmax)) <span class="op">+</span> <span class="dv">1</span>) <span class="op">+</span> <span class="dv">1</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>rseq <span class="op">=</span> np.logspace(<span class="dv">9</span>, nseq <span class="op">-</span> <span class="dv">1</span>, num <span class="op">=</span> nseq <span class="op">-</span> <span class="dv">9</span>, base <span class="op">=</span> <span class="fl">2.0</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>rseq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="37">
<pre><code>array([  512.,  1024.,  2048.,  4096.,  8192., 16384.])</code></pre>
</div>
</div>
<div id="5cb6e40a-0293-4e68-8208-b74fcf7342ee" class="cell" data-execution_count="40">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>nnm_model <span class="op">=</span> {</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>    x: FrankWolfe(model <span class="op">=</span> <span class="st">'nnm'</span>, max_iter <span class="op">=</span> <span class="dv">10000</span>, </span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>                  svd_max_iter <span class="op">=</span> <span class="dv">50</span>, show_progress <span class="op">=</span> <span class="va">True</span>, print_skip <span class="op">=</span> <span class="dv">100</span>) <span class="cf">for</span> x <span class="kw">in</span> rseq}</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>nnm_model[<span class="dv">512</span>] <span class="op">=</span> nnm_model_</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="f844ed56-6ddf-478d-abb1-719165553952" class="cell" data-execution_count="44">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, rank <span class="kw">in</span> <span class="bu">enumerate</span>(rseq):</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>        prev_rank <span class="op">=</span> rseq[i<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>        nnm_model[rank].fit(Z_cent, rank, mask <span class="op">=</span> Z_mask, X0 <span class="op">=</span> nnm_model[prev_rank].X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>2024-01-31 14:25:22,378 | nnwmf.optimize.frankwolfe                | INFO    | Iteration 0. Step size 1.000. Duality Gap 945157
2024-01-31 14:25:59,659 | nnwmf.optimize.frankwolfe                | WARNING | Step Size is less than 0. Using last valid step size.
2024-01-31 14:26:29,154 | nnwmf.optimize.frankwolfe                | INFO    | Iteration 0. Step size 0.385. Duality Gap 1.5822e+06
2024-01-31 14:33:32,767 | nnwmf.optimize.frankwolfe                | WARNING | Step Size is less than 0. Using last valid step size.
2024-01-31 14:33:47,967 | nnwmf.optimize.frankwolfe                | WARNING | Step Size is less than 0. Using last valid step size.
2024-01-31 14:36:17,084 | nnwmf.optimize.frankwolfe                | WARNING | Step Size is less than 0. Using last valid step size.
2024-01-31 14:37:01,744 | nnwmf.optimize.frankwolfe                | WARNING | Step Size is less than 0. Using last valid step size.
2024-01-31 14:37:09,216 | nnwmf.optimize.frankwolfe                | WARNING | Step Size is less than 0. Using last valid step size.
2024-01-31 14:37:53,839 | nnwmf.optimize.frankwolfe                | WARNING | Step Size is less than 0. Using last valid step size.
2024-01-31 14:38:16,019 | nnwmf.optimize.frankwolfe                | WARNING | Step Size is less than 0. Using last valid step size.
2024-01-31 14:38:53,065 | nnwmf.optimize.frankwolfe                | INFO    | Iteration 100. Step size 0.007. Duality Gap 21721.2
2024-01-31 14:40:45,448 | nnwmf.optimize.frankwolfe                | WARNING | Step Size is less than 0. Using last valid step size.
2024-01-31 14:41:29,955 | nnwmf.optimize.frankwolfe                | WARNING | Step Size is less than 0. Using last valid step size.
2024-01-31 14:43:09,096 | nnwmf.optimize.frankwolfe                | INFO    | Iteration 0. Step size 0.167. Duality Gap 2.36519e+06
2024-01-31 14:55:55,048 | nnwmf.optimize.frankwolfe                | INFO    | Iteration 100. Step size 0.009. Duality Gap 97761.6
2024-01-31 15:08:03,494 | nnwmf.optimize.frankwolfe                | INFO    | Iteration 200. Step size 0.002. Duality Gap 33925.2
2024-01-31 15:09:30,380 | nnwmf.optimize.frankwolfe                | INFO    | Iteration 0. Step size 0.075. Duality Gap 3.62791e+06
2024-01-31 15:21:30,591 | nnwmf.optimize.frankwolfe                | INFO    | Iteration 100. Step size 0.010. Duality Gap 512949
2024-01-31 15:33:21,350 | nnwmf.optimize.frankwolfe                | INFO    | Iteration 200. Step size 0.004. Duality Gap 255372</code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>KeyboardInterrupt: </code></pre>
</div>
</div>
<div id="4f376408-e0d0-4908-9b4e-152a3ab96d20" class="cell" data-execution_count="47">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, rank <span class="kw">in</span> <span class="bu">enumerate</span>(rseq[:<span class="dv">4</span>]):</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>    nnm_rmse <span class="op">=</span> root_mean_squared_error(X_cent, nnm_model[rank].X, mask <span class="op">=</span> Z_mask)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span> (rank, nnm_rmse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>512.0 1.9266675901879784
1024.0 1.8749317132652092
2048.0 1.7887138061073196
4096.0 1.6675354537614655</code></pre>
</div>
</div>
<div id="7d44d8bd-ce50-4e71-b13a-639fb42976df" class="cell" data-execution_count="48">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure()</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>ax1 <span class="op">=</span> fig.add_subplot()</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>xvals <span class="op">=</span> rseq[:<span class="dv">4</span>]</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>yvals <span class="op">=</span> [root_mean_squared_error(X_cent, nnm_model[x].X, mask <span class="op">=</span> Z_mask) <span class="cf">for</span> x <span class="kw">in</span> xvals]</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>ax1.plot(xvals, yvals, marker <span class="op">=</span> <span class="st">'o'</span>)</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2024-01-20-panukb-data-filtering-for-masked-input_files/figure-html/cell-26-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="nnm-sparse" class="level1">
<h1>NNM-Sparse</h1>
<div id="4a286e8a-517d-4871-9749-b8d027284e39" class="cell" data-execution_count="40">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>nnm_sparse_model <span class="op">=</span> FrankWolfe(model <span class="op">=</span> <span class="st">'nnm-sparse'</span>, max_iter <span class="op">=</span> <span class="dv">10000</span>, svd_max_iter <span class="op">=</span> <span class="dv">50</span>, show_progress <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>nnm_sparse_model.fit(Z_cent, (rank_opt, <span class="fl">0.5</span>), mask <span class="op">=</span> Z_mask)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>2024-01-25 11:22:28,201 | nnwmf.optimize.frankwolfe                | INFO    | Iteration 0. Step size 0.468. Duality Gap 491221</code></pre>
</div>
</div>
<div id="b9625309-692a-4fa3-88dc-dec2c46a4fe3" class="cell" data-execution_count="43">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="st">"Nuclear Norms"</span>)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Recovered matrix (NNM): </span><span class="sc">{</span>np<span class="sc">.</span>linalg<span class="sc">.</span>norm(nnm_model.X, <span class="st">'nuc'</span>)<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Recovered matrix (NNM-Sparse): </span><span class="sc">{</span>np<span class="sc">.</span>linalg<span class="sc">.</span>norm(nnm_sparse_model.X, <span class="st">'nuc'</span>)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Nuclear Norms
Recovered matrix (NNM): 1015.0879690877024
Recovered matrix (NNM-Sparse): 1014.317846484247</code></pre>
</div>
</div>
<div id="c7930ee0-2f54-4e65-a32c-6ead5b36fbe1" class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.spatial <span class="im">import</span> procrustes</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> mean_squared_error(original, recovered, mask <span class="op">=</span> <span class="va">None</span>):</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> mask <span class="kw">is</span> <span class="va">None</span>: mask <span class="op">=</span> np.ones_like(original)</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>    n <span class="op">=</span> np.<span class="bu">sum</span>(mask)</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>    mse <span class="op">=</span> np.<span class="bu">sum</span>(np.square((original <span class="op">-</span> recovered) <span class="op">*</span> mask)) <span class="op">/</span> n</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> mse</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> peak_signal_to_noise_ratio(original, recovered, mask <span class="op">=</span> <span class="va">None</span>):</span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> mask <span class="kw">is</span> <span class="va">None</span>: mask <span class="op">=</span> np.ones_like(original)</span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>    omax <span class="op">=</span> np.<span class="bu">max</span>(original[mask <span class="op">==</span> <span class="dv">1</span>])</span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>    omin <span class="op">=</span> np.<span class="bu">min</span>(original[mask <span class="op">==</span> <span class="dv">1</span>])</span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>    maxsig2 <span class="op">=</span> np.square(omax <span class="op">-</span> omin)</span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>    mse <span class="op">=</span> mean_squared_error(original, recovered, mask)</span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a>    res <span class="op">=</span> <span class="dv">10</span> <span class="op">*</span> np.log10(maxsig2 <span class="op">/</span> mse)</span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> res</span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> matrix_dissimilarity_scores(original, recovered, mask <span class="op">=</span> <span class="va">None</span>):</span>
<span id="cb42-21"><a href="#cb42-21" aria-hidden="true" tabindex="-1"></a>    <span class="co">'''</span></span>
<span id="cb42-22"><a href="#cb42-22" aria-hidden="true" tabindex="-1"></a><span class="co">    Procrustes analysis returns the square of the Frobenius norm.</span></span>
<span id="cb42-23"><a href="#cb42-23" aria-hidden="true" tabindex="-1"></a><span class="co">    Use the rotated matrix to obtain the peak signal-to-noise ratio (PSNR).</span></span>
<span id="cb42-24"><a href="#cb42-24" aria-hidden="true" tabindex="-1"></a><span class="co">    '''</span></span>
<span id="cb42-25"><a href="#cb42-25" aria-hidden="true" tabindex="-1"></a>    k <span class="op">=</span> <span class="bu">min</span>(original.shape[<span class="dv">1</span>], recovered.shape[<span class="dv">1</span>])</span>
<span id="cb42-26"><a href="#cb42-26" aria-hidden="true" tabindex="-1"></a>    R_orig, R_recv, m2 <span class="op">=</span> procrustes(original[:, <span class="dv">0</span>:k], recovered[:, <span class="dv">0</span>:k])</span>
<span id="cb42-27"><a href="#cb42-27" aria-hidden="true" tabindex="-1"></a>    psnr <span class="op">=</span> peak_signal_to_noise_ratio(R_orig, R_recv, mask)</span>
<span id="cb42-28"><a href="#cb42-28" aria-hidden="true" tabindex="-1"></a>    rmse <span class="op">=</span> np.sqrt(mean_squared_error(R_orig, R_recv, mask))</span>
<span id="cb42-29"><a href="#cb42-29" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> rmse, psnr</span>
<span id="cb42-30"><a href="#cb42-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-31"><a href="#cb42-31" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="st">"RMSE"</span>)</span>
<span id="cb42-32"><a href="#cb42-32" aria-hidden="true" tabindex="-1"></a>orig_rmse, orig_psnr <span class="op">=</span> matrix_dissimilarity_scores(X_cent, Z_cent, Z_mask)</span>
<span id="cb42-33"><a href="#cb42-33" aria-hidden="true" tabindex="-1"></a>rpca_rmse, rpca_psnr <span class="op">=</span> matrix_dissimilarity_scores(X_cent, rpca_model.L_, Z_mask)</span>
<span id="cb42-34"><a href="#cb42-34" aria-hidden="true" tabindex="-1"></a>nnm_rmse, nnm_psnr <span class="op">=</span> matrix_dissimilarity_scores(X_cent, nnm_model[<span class="dv">4096</span>].X, Z_mask)</span>
<span id="cb42-35"><a href="#cb42-35" aria-hidden="true" tabindex="-1"></a><span class="co">#nnm_sparse_rmse, nnm_sparse_psnr = matrix_dissimilarity_scores(X_cent, nnm_sparse_model.X, Z_mask)</span></span>
<span id="cb42-36"><a href="#cb42-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-37"><a href="#cb42-37" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Zero: </span><span class="sc">{</span>orig_rmse<span class="sc">:g}</span><span class="ss">"</span>)</span>
<span id="cb42-38"><a href="#cb42-38" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"RPCA: </span><span class="sc">{</span>rpca_rmse<span class="sc">:g}</span><span class="ss">"</span>)</span>
<span id="cb42-39"><a href="#cb42-39" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"NNM: </span><span class="sc">{</span>nnm_rmse<span class="sc">:g}</span><span class="ss">"</span>)</span>
<span id="cb42-40"><a href="#cb42-40" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"NNM-Sparse: </span><span class="sc">{</span>nnm_sparse_rmse<span class="sc">:g}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>RMSE</code></pre>
</div>
</div>
<div id="9dd36a8e-6255-449f-9277-3d3ae8448c96" class="cell" data-execution_count="52">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="st">"RMSE"</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Zero prediction: </span><span class="sc">{</span>orig_rmse<span class="sc">:g}</span><span class="ss">"</span>)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"RPCA: </span><span class="sc">{</span>rpca_rmse<span class="sc">:g}</span><span class="ss">"</span>)</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"NNM: </span><span class="sc">{</span>nnm_rmse<span class="sc">:g}</span><span class="ss">"</span>)</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"NNM-Sparse: </span><span class="sc">{</span>nnm_sparse_rmse<span class="sc">:g}</span><span class="ss">"</span>)</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">""</span>)</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="st">"PSNR"</span>)</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"Zero prediction: </span><span class="sc">{</span>orig_psnr<span class="sc">:g}</span><span class="ss">"</span>)</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"RPCA: </span><span class="sc">{</span>rpca_psnr<span class="sc">:g}</span><span class="ss">"</span>)</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"NNM: </span><span class="sc">{</span>nnm_psnr<span class="sc">:g}</span><span class="ss">"</span>)</span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="ss">f"NNM-Sparse: </span><span class="sc">{</span>nnm_sparse_psnr<span class="sc">:g}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>RMSE
Zero prediction: 8.9478e-05
RPCA: 0.000446309
NNM: 0.000407213
NNM-Sparse: 0.000407142

PSNR
Zero prediction: 56.9045
RPCA: 42.9461
NNM: 43.7424
NNM-Sparse: 43.7439</code></pre>
</div>
</div>
<div id="75688c5f-369d-4cbf-807e-1f45e3133942" class="cell" data-execution_count="49">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>zero_rmse, zero_psnr <span class="op">=</span> matrix_dissimilarity_scores(X_cent, X, Z_mask)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>RMSE</code></pre>
</div>
</div>
<div id="2a4f26bf-0fa7-4549-ac46-eea768aa6b41" class="cell" data-execution_count="50">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>zero_rmse</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="50">
<pre><code>3.859249487545338e-15</code></pre>
</div>
</div>
<div id="86ebe086-97fa-480a-8738-b439f1ae94d8" class="cell" data-execution_count="51">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>zero_psnr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="51">
<pre><code>264.2087877109133</code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("^(?:http:|https:)\/\/banskt\.github\.io\/npd-notes\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
          // default icon
          link.classList.add("external");
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.dev/banskt/npd-notes/blob/main/notebooks/ukbb/2024-01-20-panukb-data-filtering-for-masked-input.ipynb" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/banskt/npd-notes/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>